// @duetds/components/lib/collection/components/duet-banner/duet-banner.js
/*!
 * Built with Duet Design System
 */
import { h, Host } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
import { getColorByName } from "../../utils/token-utils";
export class DuetBanner {
  constructor() {
    this.theme = "";
    this.margin = "auto";
    this.backgroundColor = "primary-lighter";
    this.iconColor = "gray-lightest";
    this.iconBackgroundColor = "primary";
    this.textColor = "secondary";
    this.size = "auto";
    this.image = "";
    this.icon = "";
    this.headingLevel = "h2";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
    this.hasHeadingSlot = !!this.element.querySelector('[slot="heading"]');
    this.hasDescriptionSlot = !!this.element.querySelector('[slot="description"]');
    this.hasActionSlot = !!this.element.querySelector('[slot="action"]');
  }
  /**
   * render() function
   * Always the last one of the class.
   */
  render() {
    const backgroundColorTemp = this.theme === "turva" && this.backgroundColor === "primary-lighter" ? "gray-light" : this.backgroundColor;
    const backgroundColor = getColorByName(backgroundColorTemp, this.theme);
    const bannerStyles = {
      backgroundColor,
    };
    return (h(Host, { class: { "duet-m-0": this.margin === "none" } }, h("aside", { class: {
        "duet-banner": true,
        "duet-theme-turva": this.theme === "turva",
        [this.size]: true,
      }, style: bannerStyles }, h("div", { class: "visual" }, h("div", { class: "image-container" }, h("img", { class: "image", src: this.image, alt: "", role: "presentation" })), this.icon && (h("duet-icon", { theme: this.theme, class: "icon", name: this.icon, shape: "brand", size: "xx-large", color: this.iconColor, background: this.iconBackgroundColor, "background-rotation": "7", "background-opacity": "0.85", margin: "none" }))), h("div", { class: "content" }, this.hasHeadingSlot && (h("div", { class: "heading" }, h("duet-heading", { theme: this.theme, level: this.headingLevel, "visual-level": "h2", margin: "none", color: this.textColor }, h("slot", { name: "heading" })))), this.hasDescriptionSlot && (h("div", { class: "description" }, h("duet-paragraph", { theme: this.theme, margin: "none", color: this.textColor }, h("slot", { name: "description" })))), this.hasActionSlot && (h("div", { class: "action" }, h("slot", { name: "action" })))))));
  }
  static get is() { return "duet-banner"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["duet-banner.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["duet-banner.css"]
    };
  }
  static get properties() {
    return {
      "theme": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetTheme",
          "resolved": "\"\" | \"default\" | \"turva\"",
          "references": {
            "DuetTheme": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Theme."
        },
        "attribute": "theme",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "margin": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetMargin",
          "resolved": "\"auto\" | \"none\"",
          "references": {
            "DuetMargin": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Controls the margin of the component."
        },
        "attribute": "margin",
        "reflect": false,
        "defaultValue": "\"auto\""
      },
      "backgroundColor": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetColor",
          "resolved": "string",
          "references": {
            "DuetColor": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Background color."
        },
        "attribute": "background-color",
        "reflect": false,
        "defaultValue": "\"primary-lighter\""
      },
      "iconColor": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetColor",
          "resolved": "string",
          "references": {
            "DuetColor": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Icon color."
        },
        "attribute": "icon-color",
        "reflect": false,
        "defaultValue": "\"gray-lightest\""
      },
      "iconBackgroundColor": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetColor",
          "resolved": "string",
          "references": {
            "DuetColor": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Icon background color."
        },
        "attribute": "icon-background-color",
        "reflect": false,
        "defaultValue": "\"primary\""
      },
      "textColor": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetColor",
          "resolved": "string",
          "references": {
            "DuetColor": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Text color."
        },
        "attribute": "text-color",
        "reflect": false,
        "defaultValue": "\"secondary\""
      },
      "size": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "\"auto\" | \"small\" | \"large\"",
          "resolved": "\"auto\" | \"large\" | \"small\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Size."
        },
        "attribute": "size",
        "reflect": false,
        "defaultValue": "\"auto\""
      },
      "image": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Image. Example image is 512x341;"
        },
        "attribute": "image",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "icon": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetIconName",
          "resolved": "string",
          "references": {
            "DuetIconName": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Icon."
        },
        "attribute": "icon",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "headingLevel": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetHeadingLevel",
          "resolved": "\"h1\" | \"h2\" | \"h3\" | \"h4\" | \"h5\" | \"h6\"",
          "references": {
            "DuetHeadingLevel": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "The actual heading level used in the HTML markup."
        },
        "attribute": "heading-level",
        "reflect": false,
        "defaultValue": "\"h2\""
      }
    };
  }
  static get elementRef() { return "element"; }
}
