// @duetds/components/lib/collection/components/duet-promo-card/duet-promo-card.js
/*!
 * Built with Duet Design System
 */
import { h, Host } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
import { getColorByName } from "../../utils/token-utils";
export class DuetPromoCard {
  constructor() {
    this.theme = "";
    this.margin = "auto";
    this.backgroundColor = "";
    this.textColor = "secondary";
    this.image = "";
    this.headingLevel = "h3";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
    this.hasPreHeadingSlot = !!this.element.querySelector('[slot="pre-heading"]');
    this.hasHeadingSlot = !!this.element.querySelector('[slot="heading"]');
    this.hasDescriptionSlot = !!this.element.querySelector('[slot="description"]');
    this.hasActionSlot = !!this.element.querySelector('[slot="action"]');
  }
  /**
   * render() function
   * Always the last one of the class.
   */
  render() {
    const backgroundColor = getColorByName(this.backgroundColor, this.theme);
    const bannerStyles = {
      backgroundColor: backgroundColor ? backgroundColor : "none",
    };
    return (h(Host, { class: { "duet-m-0": this.margin === "none" } }, h("aside", { class: {
        "duet-promo-card": true,
        "duet-theme-turva": this.theme === "turva",
        "no-bg-color": !this.backgroundColor,
      }, style: bannerStyles }, h("img", { class: "image", src: this.image, alt: "", role: "presentation" }), h("div", { class: "content" }, this.hasPreHeadingSlot && (h("div", { class: "pre-heading" }, h("slot", { name: "pre-heading" }))), this.hasHeadingSlot && (h("div", { class: "heading" }, h("duet-heading", { theme: this.theme, level: this.headingLevel, "visual-level": "h3", margin: "none", color: this.textColor }, h("slot", { name: "heading" })))), this.hasDescriptionSlot && (h("div", { class: "description" }, h("duet-paragraph", { theme: this.theme, margin: "none", color: this.textColor }, h("slot", { name: "description" })))), this.hasActionSlot && (h("div", { class: "action" }, h("slot", { name: "action" })))))));
  }
  static get is() { return "duet-promo-card"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["duet-promo-card.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["duet-promo-card.css"]
    };
  }
  static get properties() {
    return {
      "theme": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetTheme",
          "resolved": "\"\" | \"default\" | \"turva\"",
          "references": {
            "DuetTheme": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Theme."
        },
        "attribute": "theme",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "margin": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetMargin",
          "resolved": "\"auto\" | \"none\"",
          "references": {
            "DuetMargin": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Controls the margin of the component."
        },
        "attribute": "margin",
        "reflect": false,
        "defaultValue": "\"auto\""
      },
      "backgroundColor": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetColor | \"\"",
          "resolved": "string",
          "references": {
            "DuetColor": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Background color."
        },
        "attribute": "background-color",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "textColor": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetColor",
          "resolved": "string",
          "references": {
            "DuetColor": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Text color."
        },
        "attribute": "text-color",
        "reflect": false,
        "defaultValue": "\"secondary\""
      },
      "image": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Image. Example image is 512x341;"
        },
        "attribute": "image",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "headingLevel": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetHeadingLevel",
          "resolved": "\"h1\" | \"h2\" | \"h3\" | \"h4\" | \"h5\" | \"h6\"",
          "references": {
            "DuetHeadingLevel": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "The actual heading level used in the HTML markup."
        },
        "attribute": "heading-level",
        "reflect": false,
        "defaultValue": "\"h3\""
      }
    };
  }
  static get elementRef() { return "element"; }
}
