// @duetds/components/lib/collection/components/duet-page-heading/duet-page-heading.js
/*!
 * Built with Duet Design System
 */
import { h } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
/**
 * @slot heading - This is a slot for heading content.
 */
export class DuetPageHeading {
  constructor() {
    this.theme = "";
    this.icon = "";
    this.layout = "narrow";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h("div", { class: {
        "duet-page-heading": true,
        "duet-theme-turva": this.theme === "turva",
        [this.layout]: true,
      } }, h("div", { class: "background" }, this.theme === "turva" && h("div", { class: "background-mask" }), this.theme !== "turva" && h("div", { class: "brand-shape" }), this.theme !== "turva" && h("div", { class: "brand-shape second-shape" }), this.icon && (h("div", { class: "brand-shaped-icon" }, h("duet-icon", { background: "none", name: this.icon, margin: "none", size: "auto", responsive: false })))), h("div", { class: "heading" }, h("slot", { name: "heading" }))));
  }
  static get is() { return "duet-page-heading"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["duet-page-heading.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["duet-page-heading.css"]
    };
  }
  static get properties() {
    return {
      "theme": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetTheme",
          "resolved": "\"\" | \"default\" | \"turva\"",
          "references": {
            "DuetTheme": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Theme of the hero area."
        },
        "attribute": "theme",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "icon": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetIconName",
          "resolved": "string",
          "references": {
            "DuetIconName": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Icon."
        },
        "attribute": "icon",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "layout": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "\"narrow\" | \"fluid\"",
          "resolved": "\"fluid\" | \"narrow\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Layout."
        },
        "attribute": "layout",
        "reflect": false,
        "defaultValue": "\"narrow\""
      }
    };
  }
  static get elementRef() { return "element"; }
}
