// @duetds/components/lib/collection/components/duet-cookie-consent/duet-cookie-consent.js
/*!
 * Built with Duet Design System
 */
import { h, Host } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
export class DuetCookieConsent {
  constructor() {
    this.handleClick = e => {
      e.preventDefault();
      this.duetCookieConsent.emit({ component: "duet-cookie-consent" });
    };
    this.accessibleLabel = "Hyväksyn evästeiden käytön";
    this.theme = "";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function.
   * Always the last one in the class.
   */
  render() {
    return (h(Host, { tabindex: "0" }, h("div", { class: { "duet-cookie-consent": true, "duet-theme-turva": this.theme === "turva" } }, h("div", { class: "duet-cookie-consent-content" }, h("div", { class: "expand" }, h("slot", null)), h("duet-spacer", { size: "large", direction: "horizontal" }), h("div", null, h("duet-button", { variation: "negative", size: "small", onClick: this.handleClick, margin: "none", "aria-label": this.accessibleLabel, theme: this.theme }, "OK"))))));
  }
  static get is() { return "duet-cookie-consent"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["duet-cookie-consent.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["duet-cookie-consent.css"]
    };
  }
  static get properties() {
    return {
      "accessibleLabel": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Adds accessible label for the cookie consent ok button."
        },
        "attribute": "accessible-label",
        "reflect": false,
        "defaultValue": "\"Hyv\u00E4ksyn ev\u00E4steiden k\u00E4yt\u00F6n\""
      },
      "theme": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetTheme",
          "resolved": "\"\" | \"default\" | \"turva\"",
          "references": {
            "DuetTheme": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Theme of the cookie consent banner."
        },
        "attribute": "theme",
        "reflect": false,
        "defaultValue": "\"\""
      }
    };
  }
  static get events() {
    return [{
        "method": "duetCookieConsent",
        "name": "duetCookieConsent",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [],
          "text": "Emitted when the consent button has been clicked."
        },
        "complexType": {
          "original": "DuetCookieConsentEvent",
          "resolved": "{ component: \"duet-cookie-consent\"; }",
          "references": {
            "DuetCookieConsentEvent": {
              "location": "local",
              "path": "/Users/laaksju3/duetds/packages/components/src/components/duet-cookie-consent/duet-cookie-consent.tsx"
            }
          }
        }
      }];
  }
  static get elementRef() { return "element"; }
}
