// @duetds/components/lib/collection/components/duet-slideout-link/duet-slideout-link.js
/*!
 * Built with Duet Design System
 */
import { h, Host } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
export class DuetSlideoutLink {
  constructor() {
    this.accessibleLabel = undefined;
    this.theme = "";
    this.targetLanguage = undefined;
    this.active = false;
    this.href = "#";
    this.icon = undefined;
    this.size = "inherit";
    this.caret = false;
    this.menuItem = false;
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, null, h("a", { class: {
        "duet-slideout-link": true,
        "duet-theme-turva": this.theme === "turva",
        active: this.active,
        [this.size]: true,
      }, href: this.href, "aria-current": this.active ? "true" : "false", lang: this.targetLanguage, hreflang: this.targetLanguage, "aria-label": this.accessibleLabel }, this.icon && (h("duet-icon", { name: this.icon, color: "secondary", size: this.menuItem ? "x-small" : "medium", margin: "none" })), h("div", { class: "label" }, h("span", null, h("slot", null), h("slot", { name: "description" })), this.caret && (h("duet-icon", { name: "action-arrow-right", color: "secondary", size: "x-small", margin: "none" }))))));
  }
  static get is() { return "duet-slideout-link"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["duet-slideout-link.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["duet-slideout-link.css"]
    };
  }
  static get properties() {
    return {
      "accessibleLabel": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Adds accessible label for the button that is only shown for screen readers.\nTypically, this label text replaces the visible text on the button for\nusers who use assistive technology."
        },
        "attribute": "accessible-label",
        "reflect": false
      },
      "theme": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetTheme",
          "resolved": "\"\" | \"default\" | \"turva\"",
          "references": {
            "DuetTheme": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Theme."
        },
        "attribute": "theme",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "targetLanguage": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "\"en\" | \"fi\" | \"sv\"",
          "resolved": "\"en\" | \"fi\" | \"sv\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Language of the link. Use this only for language selection links."
        },
        "attribute": "target-language",
        "reflect": false,
        "defaultValue": "undefined"
      },
      "active": {
        "type": "boolean",
        "mutable": true,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Active"
        },
        "attribute": "active",
        "reflect": false,
        "defaultValue": "false"
      },
      "href": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Href"
        },
        "attribute": "href",
        "reflect": false,
        "defaultValue": "\"#\""
      },
      "icon": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Icon"
        },
        "attribute": "icon",
        "reflect": true,
        "defaultValue": "undefined"
      },
      "size": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "\"inherit\" | \"small\" | \"medium\" | \"large\"",
          "resolved": "\"inherit\" | \"large\" | \"medium\" | \"small\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Size. Inherit, small, medium or large. Inherit value inherits styles from the value in the underlying slot."
        },
        "attribute": "size",
        "reflect": true,
        "defaultValue": "\"inherit\""
      },
      "caret": {
        "type": "boolean",
        "mutable": true,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Caret"
        },
        "attribute": "caret",
        "reflect": false,
        "defaultValue": "false"
      },
      "menuItem": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "internal",
              "text": "Indicates the component is used where aria-role menuitem is expected"
            }],
          "text": ""
        },
        "attribute": "menu-item",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
  static get elementRef() { return "element"; }
}
