// @duetds/components/lib/collection/components/duet-date-picker/date-i18n.js
/*!
 * Built with Duet Design System
 */
const formatOptionsShort = { day: "numeric", month: "long" };
const formatOptionsLong = { day: "numeric", month: "long", year: "numeric" };
const i18n = {
  fi: {
    buttonLabel: "Valitse päivämäärä",
    prevMonthLabel: "Edellinen kuukausi",
    nextMonthLabel: "Seuraava kuukausi",
    monthSelectLabel: "Kuukausi",
    yearSelectLabel: "Vuosi",
    closeLabel: "Sulje ikkuna",
    selected: "Valittu päivämäärä on",
    keyboardInstruction: "Voit navigoida päivämääriä nuolinäppäimillä",
    dayLabels: ["Maanantai", "Tiistai", "Keskiviikko", "Torstai", "Perjantai", "Lauantai", "Sunnuntai"],
    monthLabels: [
      "Tammikuu",
      "Helmikuu",
      "Maaliskuu",
      "Huhtikuu",
      "Toukokuu",
      "Kesäkuu",
      "Heinäkuu",
      "Elokuu",
      "Syyskuu",
      "Lokakuu",
      "Marraskuu",
      "Joulukuu",
    ],
    monthLabelsShort: [
      "Tammi",
      "Helmi",
      "Maalis",
      "Huhti",
      "Touko",
      "Kesä",
      "Heinä",
      "Elo",
      "Syys",
      "Loka",
      "Marras",
      "Joulu",
    ],
    formatterShort: new Intl.DateTimeFormat("fi-FI", formatOptionsShort),
    formatterLong: new Intl.DateTimeFormat("fi-FI", formatOptionsLong),
  },
  en: {
    buttonLabel: "Choose date",
    prevMonthLabel: "Previous month",
    nextMonthLabel: "Next month",
    monthSelectLabel: "Month",
    yearSelectLabel: "Year",
    closeLabel: "Close window",
    selected: "Selected date is",
    keyboardInstruction: "You can use arrow keys to navigate dates",
    dayLabels: ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"],
    monthLabels: [
      "January",
      "February",
      "March",
      "April",
      "May",
      "June",
      "July",
      "August",
      "September",
      "October",
      "November",
      "December",
    ],
    monthLabelsShort: ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"],
    formatterShort: new Intl.DateTimeFormat("en-GB", formatOptionsShort),
    formatterLong: new Intl.DateTimeFormat("en-GB", formatOptionsLong),
  },
  sv: {
    buttonLabel: "Välj datum",
    prevMonthLabel: "Föregående månad",
    nextMonthLabel: "Nästa månad",
    monthSelectLabel: "Månad",
    yearSelectLabel: "År",
    closeLabel: "Stäng fönstret",
    selected: "Valt datum är",
    keyboardInstruction: "Använd piltangenterna för att navigera i kalender",
    dayLabels: ["Måndag", "Tisdag", "Onsdag", "Torsdag", "Fredag", "Lördag", "Söndag"],
    monthLabels: [
      "Januari",
      "Februari",
      "Mars",
      "April",
      "Maj",
      "Juni",
      "Juli",
      "Augusti",
      "September",
      "Oktober",
      "November",
      "December",
    ],
    monthLabelsShort: ["Jan", "Feb", "Mars", "April", "Maj", "Juni", "Juli", "Aug", "Sep", "Okt", "Nov", "Dec"],
    formatterShort: new Intl.DateTimeFormat("sv-SE", formatOptionsShort),
    formatterLong: new Intl.DateTimeFormat("sv-SE", formatOptionsLong),
  },
};
export default i18n;
