// @duetds/components/lib/collection/components/duet-heading/duet-heading.js
/*!
 * Built with Duet Design System
 */
import { h } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
import { sanitizeString } from "../../utils/string-utils";
import { getColorByName } from "../../utils/token-utils";
/**
 * @part duet-heading - piercing selector for styling the heading element that should be used only in special circumstances
 */
export class DuetHeading {
  constructor() {
    this.level = "h2";
    this.margin = "auto";
    this.visualLevel = undefined;
    this.border = false;
    this.weight = "auto";
    this.color = "";
    this.theme = "";
    this.hyphenate = false;
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    const TagName = sanitizeString(this.level);
    const color = getColorByName(this.color);
    return (h(TagName, { class: {
        "duet-heading": true,
        "duet-heading-border": this.border !== false,
        "duet-heading-border-solid": this.border === "solid",
        "duet-normal": this.weight === "normal",
        "duet-semibold": this.weight === "semibold",
        "duet-heading-hyphenate": this.hyphenate,
        "duet-m-0": this.margin === "none",
        [this.visualLevel ? this.visualLevel : this.level]: true,
        "duet-theme-turva": this.theme === "turva",
      }, style: { color }, part: "duet-heading" }, h("slot", null)));
  }
  static get is() { return "duet-heading"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["duet-heading.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["duet-heading.css"]
    };
  }
  static get properties() {
    return {
      "level": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetHeadingLevel",
          "resolved": "\"h1\" | \"h2\" | \"h3\" | \"h4\" | \"h5\" | \"h6\"",
          "references": {
            "DuetHeadingLevel": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "The actual heading level used in the HTML markup."
        },
        "attribute": "level",
        "reflect": true,
        "defaultValue": "\"h2\""
      },
      "margin": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetMargin",
          "resolved": "\"auto\" | \"none\"",
          "references": {
            "DuetMargin": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Controls the margin of the component."
        },
        "attribute": "margin",
        "reflect": false,
        "defaultValue": "\"auto\""
      },
      "visualLevel": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetVisualHeadingLevel",
          "resolved": "\"h0\" | \"h1\" | \"h2\" | \"h3\" | \"h4\" | \"h5\" | \"h6\"",
          "references": {
            "DuetVisualHeadingLevel": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Make the visual style mimic a specific heading level. This option allows\nyou to make e.g. h1 visually look like h3, but still keep it h1 in the\nmarkup."
        },
        "attribute": "visual-level",
        "reflect": false
      },
      "border": {
        "type": "any",
        "mutable": false,
        "complexType": {
          "original": "boolean | \"solid\"",
          "resolved": "\"solid\" | boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Enable or disable the border underneath the heading, solid will darken the border"
        },
        "attribute": "border",
        "reflect": false,
        "defaultValue": "false"
      },
      "weight": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetHeadingWeight",
          "resolved": "\"auto\" | \"normal\" | \"semibold\"",
          "references": {
            "DuetHeadingWeight": {
              "location": "local",
              "path": "/Users/laaksju3/duetds/packages/components/src/components/duet-heading/duet-heading.tsx"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "This property allows you to set normal or semibold font weight for all sizes\nwhen needed."
        },
        "attribute": "weight",
        "reflect": false,
        "defaultValue": "\"auto\""
      },
      "color": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetColor",
          "resolved": "string",
          "references": {
            "DuetColor": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Custom color for the heading as a design token entered in camelCase or\nkebab-case. Example: \"color-primary\"."
        },
        "attribute": "color",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "theme": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetTheme",
          "resolved": "\"\" | \"default\" | \"turva\"",
          "references": {
            "DuetTheme": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Theme of the heading."
        },
        "attribute": "theme",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "hyphenate": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Enable hyphenation for long titles. Useful when there is no control over content e.g. user-generated content."
        },
        "attribute": "hyphenate",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
  static get elementRef() { return "element"; }
}
