// @duetds/components/lib/collection/utils/keyboard-utils.js
/*!
 * Built with Duet Design System
 */
export const isArrowLeftKey = (e) => e.keyCode === 37;
export const isArrowUpKey = (e) => e.keyCode === 38;
export const isArrowRightKey = (e) => e.keyCode === 39;
export const isArrowDownKey = (e) => e.keyCode === 40;
export const isArrowKey = (e) => isArrowDownKey(e) || isArrowRightKey(e) || isArrowUpKey(e) || isArrowLeftKey(e);
export const isEscapeKey = (e) => e.key === "Escape" || e.key === "Esc" || e.keyCode === 27;
export const isTabKey = (e) => e.key === "Tab" || e.keyCode === 9;
export const isBackspaceKey = (e) => e.keyCode === 8;
export const isSpaceKey = (e) => e.key === "Space" || e.keyCode === 32;
export const isPlusKey = (e) => e.key === "+" || e.keyCode === 43;
export const isMinusKey = (e) => e.key === "-" || e.keyCode === 45;
export const isQuestionKey = (e) => e.key === "?";
export const isEnterKey = (e) => e.key === "Enter" || e.keyCode === 13;
export const isKeyboardClick = (e) => isEnterKey(e) || isSpaceKey(e);
export const isNumber = (e) => (e.keyCode >= 48 && e.keyCode <= 57) || (e.keyCode >= 96 && e.keyCode <= 105);
