// @duetds/components/lib/collection/components/duet-divider/duet-divider.js
/*!
 * Built with Duet Design System
 */
import { h } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
import { getColorByName } from "../../utils/token-utils";
export class DuetDivider {
  constructor() {
    this.theme = "";
    this.margin = "auto";
    this.color = "";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    const background = getColorByName(this.color);
    return (h("div", { class: {
        "duet-m-0": this.margin === "none",
        "duet-m-s": this.margin === "small",
        "duet-divider": true,
        "duet-theme-turva": this.theme === "turva",
      }, style: { background } }));
  }
  static get is() { return "duet-divider"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["duet-divider.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["duet-divider.css"]
    };
  }
  static get properties() {
    return {
      "theme": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetTheme",
          "resolved": "\"\" | \"default\" | \"turva\"",
          "references": {
            "DuetTheme": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Theme of the divider."
        },
        "attribute": "theme",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "margin": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetDividerMargin",
          "resolved": "\"auto\" | \"none\" | \"small\"",
          "references": {
            "DuetDividerMargin": {
              "location": "local",
              "path": "/Users/laaksju3/duetds/packages/components/src/components/duet-divider/duet-divider.tsx"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Controls the margin of the component."
        },
        "attribute": "margin",
        "reflect": false,
        "defaultValue": "\"auto\""
      },
      "color": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetColor",
          "resolved": "string",
          "references": {
            "DuetColor": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Custom color, as a design token entered in camelCase or kebab-case.\nExample: \"color-primary\"."
        },
        "attribute": "color",
        "reflect": false,
        "defaultValue": "\"\""
      }
    };
  }
  static get elementRef() { return "element"; }
}
