// @duetds/components/lib/collection/components/duet-empty-state/duet-empty-state.js
/*!
 * Built with Duet Design System
 */
import { h, Host } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
export class DuetEmptyState {
  constructor() {
    this.theme = "";
    this.icon = "messaging-question";
    this.size = "medium";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, null, h("div", { class: {
        "duet-empty-state": true,
        "duet-theme-turva": this.theme === "turva",
        "duet-is-small": this.size === "small",
      } }, h("duet-icon", { color: "currentColor", name: this.icon, size: this.size === "small" ? "large" : "x-large" }), h("div", null, h("slot", null)))));
  }
  static get is() { return "duet-empty-state"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["duet-empty-state.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["duet-empty-state.css"]
    };
  }
  static get properties() {
    return {
      "theme": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetTheme",
          "resolved": "\"\" | \"default\" | \"turva\"",
          "references": {
            "DuetTheme": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Theme of the empty state component."
        },
        "attribute": "theme",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "icon": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetIconName",
          "resolved": "string",
          "references": {
            "DuetIconName": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Icon shown in the empty state component."
        },
        "attribute": "icon",
        "reflect": false,
        "defaultValue": "\"messaging-question\""
      },
      "size": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetEmptyStateSize",
          "resolved": "\"medium\" | \"small\"",
          "references": {
            "DuetEmptyStateSize": {
              "location": "local",
              "path": "/Users/laaksju3/duetds/packages/components/src/components/duet-empty-state/duet-empty-state.tsx"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "The size of the empty state message."
        },
        "attribute": "size",
        "reflect": false,
        "defaultValue": "\"medium\""
      }
    };
  }
}
