// @duetds/components/lib/collection/components/duet-nav/duet-nav.js
/*!
 * Built with Duet Design System
 */
import { h, Host } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
/**
 * @slot default - Slot for keyboard accessible initially hidden links. Use duet-link with tab-to-link attribute.
 * @slot mobile - Slot for navigation in small viewports (less than 992px wide).
 * @slot desktop - Slot for navigation in large viewports (992px wide or more).
 */
export class DuetNav {
  constructor() {
    this.theme = "";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, null, h("div", null, h("slot", null), h("slot", { name: "mobile" }), h("slot", { name: "desktop" }))));
  }
  static get is() { return "duet-nav"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["duet-nav.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["duet-nav.css"]
    };
  }
  static get properties() {
    return {
      "theme": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetTheme",
          "resolved": "\"\" | \"default\" | \"turva\"",
          "references": {
            "DuetTheme": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Theme."
        },
        "attribute": "theme",
        "reflect": false,
        "defaultValue": "\"\""
      }
    };
  }
  static get elementRef() { return "element"; }
}
