// @duetds/components/lib/esm/app-globals-de2a5510.js
/*!
 * Built with Duet Design System
 */
const setupLoggers = () => {
  {
    globalThis.ds = {
      log: () => {
        return;
      },
      warn: () => {
        return;
      },
      info: () => {
        return;
      },
    };
  }
};

const TAB_KEY = 9;
const appGlobalScript = () => {
  function listenKeys(event) {
    if (event.keyCode === TAB_KEY) {
      var components = document.querySelectorAll(".hydrated");
      for (var i = 0; i < components.length; i++) {
        components[i].classList.add("user-is-tabbing");
      }
      if (typeof window !== "undefined") {
        window.removeEventListener("keydown", listenKeys);
        window.addEventListener("mousedown", listenMouse, false);
      }
    }
  }
  function listenMouse() {
    var components = document.querySelectorAll(".hydrated");
    for (var i = 0; i < components.length; i++) {
      components[i].classList.remove("user-is-tabbing");
    }
    if (typeof window !== "undefined") {
      window.removeEventListener("mousedown", listenMouse);
      window.addEventListener("keydown", listenKeys, false);
    }
  }
  if (typeof window !== "undefined") {
    window.addEventListener("keydown", listenKeys, false);
  }
  setupLoggers();
};

const globalScripts = appGlobalScript;

export { globalScripts as g };
