// @duetds/components/lib/esm/duet-menu-bar-dropdown-link.entry.js
/*!
 * Built with Duet Design System
 */
import { r as registerInstance, h, H as Host, g as getElement } from './index-b45ea754.js';
import { i as inheritGlobalTheme } from './themeable-component-cb76be7c.js';
import { f as isEscapeKey } from './keyboard-utils-d237bd5a.js';

const duetMenuBarDropdownLinkCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;width:100%;height:100%}.duet-menu-bar-dropdown-link{display:flex;flex-direction:row;gap:0.5rem;align-items:center;justify-content:flex-start;height:100%;padding:0.5rem 20px;font-family:\"localtapiola-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";font-size:0.875rem;font-weight:400;line-height:1.25;color:rgb(0, 41, 77);text-align:center;text-decoration:none}.duet-menu-bar-dropdown-link.duet-theme-turva{font-family:\"turva-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";color:rgb(23, 28, 58)}.duet-menu-bar-dropdown-link--slot-right:hover{background:rgb(230, 242, 248)}.duet-menu-bar-dropdown-link--slot-right:hover.duet-theme-turva{background:rgb(228, 228, 230)}.duet-menu-bar-dropdown-link--slot-right:focus{outline:0;box-shadow:0 0 0 4px rgba(0, 119, 179, 0.3), 0 0 0 1px rgb(0, 119, 179)}.duet-menu-bar-dropdown-link--slot-right:focus.duet-theme-turva{box-shadow:0 0 0 4px rgba(23, 28, 58, 0.2), 0 0 0 1px rgb(23, 28, 58)}.duet-menu-bar-dropdown-link--slot-below:hover{color:rgb(0, 77, 128);background:rgb(243, 249, 252)}.duet-menu-bar-dropdown-link--slot-below:hover.duet-theme-turva{color:rgb(148, 9, 37);background:rgb(228, 228, 230)}";

const DuetMenuBarDropdownLink = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.getVariant = () => {
      const slot = this.element.parentElement.parentElement.getAttribute("slot");
      switch (slot) {
        case "center":
          return "right";
        case "right":
          return "right";
        default:
          return "right";
      }
    };
    this.theme = "";
    this.href = "#";
    this.icon = undefined;
    this.accessibleLabel = undefined;
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  handleEscape(e) {
    if (isEscapeKey(e)) {
      const dropdown = this.element.closest("duet-menu-bar-dropdown");
      dropdown.open = false;
      dropdown.setFocus();
    }
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, null, h("a", { class: {
        "duet-menu-bar-dropdown-link": true,
        [`duet-menu-bar-dropdown-link--slot-${this.getVariant()}`]: true,
        "duet-theme-turva": this.theme === "turva",
      }, href: this.href, role: "menuitem", onKeyUp: e => this.handleEscape(e), "aria-label": this.accessibleLabel }, this.icon && (h("duet-icon", { theme: this.theme, name: this.icon, size: "x-small", margin: "none", color: "currentColor" })), h("slot", null))));
  }
  get element() { return getElement(this); }
};
DuetMenuBarDropdownLink.style = duetMenuBarDropdownLinkCss;

export { DuetMenuBarDropdownLink as duet_menu_bar_dropdown_link };
