// @duetds/components/lib/esm/duet-breadcrumbs.entry.js
/*!
 * Built with Duet Design System
 */
import { r as registerInstance, h, F as Fragment, H as Host, g as getElement } from './index-b45ea754.js';
import { i as inheritGlobalTheme } from './themeable-component-cb76be7c.js';
import { a as getLocaleString } from './language-utils-6be16ff0.js';
import './string-utils-5954f458.js';

var actionArrowRightSmall={"title":"action-arrow-right-small","tags":"action arrow right small caret","svg":"<svg fill=\"currentColor\" viewBox=\"0 0 24 24\" xmlns=\"http://www.w3.org/2000/svg\" role=\"img\"><path d=\"M19.5 12a2.3 2.3 0 0 1-.78 1.729L7.568 23.54a1.847 1.847 0 0 1-2.439-2.773l9.752-8.579a.25.25 0 0 0 0-.376L5.129 3.233A1.847 1.847 0 0 1 7.568.46l11.148 9.808A2.31 2.31 0 0 1 19.5 12z\"/></svg>"};

const duetBreadcrumbsCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:block;width:100%}.duet-breadcrumbs{position:relative;display:flex;align-items:center;justify-content:center;font-family:\"localtapiola-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\"}.duet-breadcrumbs .container{position:relative;display:flex;align-items:center;justify-content:space-between;width:100%;max-width:1384px;min-height:36px;margin:0 28px}.duet-breadcrumbs ol{display:flex;flex-direction:row;flex-wrap:wrap;gap:8px;align-items:center;justify-content:flex-start;padding:10px 0;color:inherit;list-style:none}.duet-breadcrumbs ol li{font-weight:400}.duet-breadcrumbs ol li:last-of-type{font-weight:600}.duet-breadcrumbs .duet-breadrumbs-arrow{display:inherit;align-items:center;justify-content:center;width:0.375rem;height:0.375rem}.duet-breadcrumbs .hide{display:none}.duet-breadcrumbs .bottom-border{position:absolute;bottom:0;display:none;width:100%;height:1px}.duet-breadcrumbs.header{color:rgb(0, 41, 77);background:rgb(243, 249, 252)}.duet-breadcrumbs.header ol duet-breadcrumb[href]:hover{color:rgb(0, 77, 128)}.duet-breadcrumbs.header .bottom-border--nav{display:block;background:rgb(225, 227, 230)}.duet-breadcrumbs.header.duet-theme-turva{color:rgb(23, 28, 58);background:rgb(245, 245, 247)}.duet-breadcrumbs.header.duet-theme-turva ol duet-breadcrumb[href]:hover{color:rgb(148, 9, 37)}.duet-breadcrumbs.footer{color:rgb(255, 255, 255);background:rgb(0, 41, 77)}@media (min-width: 64.0625em){.duet-breadcrumbs.footer{justify-content:flex-start}}@media (min-width: 76.25em){.duet-breadcrumbs.footer{justify-content:center}}.duet-breadcrumbs.footer .container{max-width:1110px}@media (min-width: 62em){.duet-breadcrumbs.footer .container{margin:0 72px 0 48px}}.duet-breadcrumbs.footer ol{padding:20px 0}@media (min-width: 62em){.duet-breadcrumbs.footer ol{padding:20px 8px}}.duet-breadcrumbs.footer ol duet-breadcrumb[href]:hover{text-decoration:underline}.duet-breadcrumbs.footer .bottom-border--container{display:block;background:rgba(255, 255, 255, 0.2)}.duet-breadcrumbs.footer.duet-theme-turva{color:rgb(255, 255, 255);background:rgb(198, 12, 48)}";

const DuetBreadcrumbs = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.observers = [];
    this.breadcrumbsAriaLabel = getLocaleString({
      fi: "Murupolku",
      sv: "Ströbröd",
      en: "Breadcrumbs",
    });
    this.children = undefined;
    this.theme = "";
    this.variation = "header";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
    this.children = Array.from(this.element.children);
  }
  connectedCallback() {
    Array.from(this.element.children).forEach(child => {
      const observer = new MutationObserver(() => (this.children = Array.from(this.element.children)));
      observer.observe(child, {
        childList: true,
        subtree: true,
        attributes: true,
        characterData: true,
      });
      this.observers = [...this.observers, observer];
    });
  }
  disconnectedCallback() {
    this.observers.forEach(observer => observer.disconnect());
    this.observers = [];
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, null, h("nav", { "aria-label": this.breadcrumbsAriaLabel, class: {
        "duet-breadcrumbs": true,
        "duet-theme-turva": this.theme === "turva",
        [this.variation]: true,
      } }, h("div", { class: "container" }, h("ol", null, this.children.map((child, index) => (h(Fragment, null, index > 0 && (h("li", { "aria-hidden": "true" }, h("duet-icon", { class: "duet-breadrumbs-arrow", size: "auto", icon: actionArrowRightSmall.svg, margin: "none", color: "currentColor" }))), h("li", { innerHTML: child.outerHTML }), h("div", { class: "hide" }, h("slot", null)))))), h("div", { class: "bottom-border bottom-border--container" })), h("div", { class: "bottom-border bottom-border--nav" }))));
  }
  get element() { return getElement(this); }
};
DuetBreadcrumbs.style = duetBreadcrumbsCss;

export { DuetBreadcrumbs as duet_breadcrumbs };
