// @duetds/components/lib/esm/duet-toolbar-dropdown-link.entry.js
/*!
 * Built with Duet Design System
 */
import { r as registerInstance, h, H as Host, g as getElement } from './index-b45ea754.js';
import { i as inheritGlobalTheme } from './themeable-component-cb76be7c.js';
import { f as isEscapeKey } from './keyboard-utils-d237bd5a.js';

const duetToolbarDropdownLinkCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;width:100%;height:100%}.duet-toolbar-dropdown-link{display:flex;flex-direction:row;gap:0.5rem;align-items:center;justify-content:flex-start;height:100%;padding:0.5rem 20px;font-family:\"localtapiola-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";font-size:0.875rem;font-weight:400;line-height:1.25;color:rgb(0, 41, 77);text-align:center;text-decoration:none}.duet-toolbar-dropdown-link.duet-theme-turva{font-family:\"turva-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";color:rgb(23, 28, 58)}.duet-toolbar-dropdown-link:focus{outline:0;box-shadow:0 0 0 4px rgba(0, 119, 179, 0.3), 0 0 0 1px rgb(0, 119, 179)}.duet-toolbar-dropdown-link:focus.duet-theme-turva{box-shadow:0 0 0 4px rgba(23, 28, 58, 0.2), 0 0 0 1px rgb(23, 28, 58)}.duet-toolbar-dropdown-link.active,.duet-toolbar-dropdown-link.active:hover{color:rgb(255, 255, 255);background:rgb(0, 119, 179)}.duet-toolbar-dropdown-link.active.duet-theme-turva,.duet-toolbar-dropdown-link.active:hover.duet-theme-turva{color:rgb(255, 255, 255);background:rgb(198, 12, 48)}.duet-toolbar-dropdown-link:hover{background:rgb(230, 242, 248)}.duet-toolbar-dropdown-link:hover.duet-theme-turva{background:rgb(228, 228, 230)}";

const DuetToolbarDropdownLink = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.accessibleLabel = undefined;
    this.theme = "";
    this.targetLanguage = undefined;
    this.href = "#";
    this.icon = undefined;
    this.active = false;
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  handleEscape(e) {
    if (isEscapeKey(e)) {
      const dropdown = this.element.closest("duet-toolbar-dropdown");
      dropdown.open = false;
      dropdown.setFocus();
    }
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, null, h("a", { class: {
        "duet-toolbar-dropdown-link": true,
        "duet-theme-turva": this.theme === "turva",
        active: this.active,
      }, href: this.href, lang: this.targetLanguage, hreflang: this.targetLanguage, role: "menuitem", "aria-label": this.accessibleLabel, "aria-current": this.active ? "true" : "false", onKeyUp: e => this.handleEscape(e) }, this.icon && h("duet-icon", { name: this.icon, size: "x-small", margin: "none" }), h("slot", null))));
  }
  get element() { return getElement(this); }
};
DuetToolbarDropdownLink.style = duetToolbarDropdownLinkCss;

export { DuetToolbarDropdownLink as duet_toolbar_dropdown_link };
