// @duetds/components/lib/esm/duet-checkmark.entry.js
/*!
 * Built with Duet Design System
 */
import { r as registerInstance, h, H as Host } from './index-b45ea754.js';
import { i as inheritGlobalTheme } from './themeable-component-cb76be7c.js';

const duetCheckmarkCss = ".duet-checkmark-container{position:relative;width:20px;height:20px;margin-left:auto;border:1px solid rgb(144, 149, 153);border-radius:4px;transition:border 300ms ease}.duet-checkmark-container.center{margin-right:auto}.duet-theme-turva .duet-checkmark-container{border-color:rgb(116, 116, 117)}.checked .duet-checkmark-container{background:rgb(0, 119, 179);border-color:rgb(0, 119, 179);transition:none}.duet-theme-turva .checked .duet-checkmark-container{background:rgb(23, 28, 58);border-color:rgb(23, 28, 58) !important}.duet-checkmark-container.duet-checkmark-radio{border-radius:50%}[disabled]+label .duet-checkmark-container{border-color:rgb(207, 210, 212) !important}.duet-theme-turva [disabled]+label .duet-checkmark-container{border-color:rgb(207, 207, 209) !important}.checked [disabled]+label .duet-checkmark-container{background:rgb(207, 210, 212) !important}.duet-theme-turva .checked [disabled]+label .duet-checkmark-container{background:rgb(207, 207, 209) !important}.duet-checkmark{position:absolute;top:10px;left:3px;width:5.3333333333px;height:11.8518518519px;border-top:2px solid rgb(255, 255, 255);border-right:2px solid rgb(255, 255, 255);opacity:1;transform:scaleX(-1) rotate(135deg) translateZ(0);transform-origin:left top;animation-name:duet-checkmark;animation-duration:600ms;animation-timing-function:ease-out;animation-fill-mode:forwards}@keyframes duet-checkmark{0%{width:0;height:0;opacity:1}20%{width:5.3333333333px;height:0;opacity:1}40%,100%{width:5.3333333333px;height:11.8518518519px;opacity:1}}";

const DuetCheckmark = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.theme = "";
    this.checked = false;
    this.presentationOnly = false;
    this.center = false;
    this.type = "checkbox";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  render() {
    return (h(Host, { class: { "duet-theme-turva": this.theme === "turva" } }, h("div", { role: this.presentationOnly ? "presentation" : null, class: { checked: this.checked } }, h("div", { class: {
        "duet-checkmark-container": true,
        center: this.center,
        "duet-checkmark-radio": this.type === "radio",
      } }, this.checked && h("div", { class: "duet-checkmark" })))));
  }
};
DuetCheckmark.style = duetCheckmarkCss;

export { DuetCheckmark as duet_checkmark };
