// @duetds/components/lib/cjs/duet-promo-card.cjs.entry.js
/*!
 * Built with Duet Design System
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-a6cfed65.js');
const themeableComponent = require('./themeable-component-fdeaef9f.js');
const tokenUtils = require('./token-utils-c1864408.js');
require('./tokens.module-5fe478ed.js');

const duetPromoCardCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;margin-right:4px !important;margin-bottom:2px !important;display:block;width:auto}:host:last-child,:host:last-of-type{margin-right:0 !important}:host(.duet-m-0){margin:0 !important}.duet-promo-card{display:flex;flex-direction:column;width:100%;height:100%;font-family:\"localtapiola-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";font-style:normal;color:rgb(0, 41, 77);text-align:left;border-radius:20px}.duet-promo-card.duet-p-0{padding:0 !important}.duet-promo-card.duet-m-0{margin:0 !important}.duet-promo-card .image{position:relative;width:100%;overflow:hidden;border-top-left-radius:20px;border-top-right-radius:20px;object-fit:cover}.duet-promo-card .content{display:flex;flex-direction:column;flex-grow:1;gap:8px;padding:20px}@media (min-width: 36em){.duet-promo-card .content{padding:28px}}.duet-promo-card .action{display:flex;flex-grow:1;align-items:flex-end;padding-top:16px}.duet-promo-card.no-bg-color .content{padding:20px 0 0}.duet-promo-card.no-bg-color .image{border-radius:16px}.duet-promo-card.duet-theme-turva{font-family:\"turva-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\"}";

const DuetPromoCard = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.theme = "";
    this.margin = "auto";
    this.backgroundColor = "";
    this.textColor = "secondary";
    this.image = "";
    this.headingLevel = "h3";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    themeableComponent.inheritGlobalTheme(this);
    this.hasPreHeadingSlot = !!this.element.querySelector('[slot="pre-heading"]');
    this.hasHeadingSlot = !!this.element.querySelector('[slot="heading"]');
    this.hasDescriptionSlot = !!this.element.querySelector('[slot="description"]');
    this.hasActionSlot = !!this.element.querySelector('[slot="action"]');
  }
  /**
   * render() function
   * Always the last one of the class.
   */
  render() {
    const backgroundColor = tokenUtils.getColorByName(this.backgroundColor, this.theme);
    const bannerStyles = {
      backgroundColor: backgroundColor ? backgroundColor : "none",
    };
    return (index.h(index.Host, { class: { "duet-m-0": this.margin === "none" } }, index.h("aside", { class: {
        "duet-promo-card": true,
        "duet-theme-turva": this.theme === "turva",
        "no-bg-color": !this.backgroundColor,
      }, style: bannerStyles }, index.h("img", { class: "image", src: this.image, alt: "", role: "presentation" }), index.h("div", { class: "content" }, this.hasPreHeadingSlot && (index.h("div", { class: "pre-heading" }, index.h("slot", { name: "pre-heading" }))), this.hasHeadingSlot && (index.h("div", { class: "heading" }, index.h("duet-heading", { theme: this.theme, level: this.headingLevel, "visual-level": "h3", margin: "none", color: this.textColor }, index.h("slot", { name: "heading" })))), this.hasDescriptionSlot && (index.h("div", { class: "description" }, index.h("duet-paragraph", { theme: this.theme, margin: "none", color: this.textColor }, index.h("slot", { name: "description" })))), this.hasActionSlot && (index.h("div", { class: "action" }, index.h("slot", { name: "action" })))))));
  }
  get element() { return index.getElement(this); }
};
DuetPromoCard.style = duetPromoCardCss;

exports.duet_promo_card = DuetPromoCard;
