// @duetds/components/lib/cjs/duet-slideout-lang.cjs.entry.js
/*!
 * Built with Duet Design System
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-a6cfed65.js');
const themeableComponent = require('./themeable-component-fdeaef9f.js');
const languageUtils = require('./language-utils-44df1d37.js');
require('./string-utils-5cadd058.js');

const duetSlideoutLangCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}.duet-slideout-lang{display:flex;gap:20px;align-items:center;justify-content:flex-start;width:min-content}::slotted(*){position:relative}::slotted(:not(:first-child)){padding-right:0;padding-left:0.5rem}::slotted(:not(:last-child))::after{position:absolute;top:15%;right:-16px;font-family:\"localtapiola-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";font-size:16px;font-weight:600;line-height:150%;color:rgb(0, 41, 77);content:\"|\"}.duet-theme-turva ::slotted(:not(:last-child))::after{color:rgb(23, 28, 58)}";

const DuetSlideoutLang = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * Own Properties.
     */
    this.chooseLangObject = {
      en: "Choose language",
      fi: "Valitse kieli",
      sv: "Välj språk",
    };
    this.language = languageUtils.getLanguage();
    this.theme = "";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    themeableComponent.inheritGlobalTheme(this);
  }
  connectedCallback() {
    languageUtils.connectLanguageChangeObserver(this);
  }
  disconnectedCallback() {
    languageUtils.disconnectLanguageChangeObserver(this);
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    const langTranslation = languageUtils.getLocaleString(this.chooseLangObject);
    return (index.h(index.Host, null, index.h("div", { class: "duet-slideout-lang", role: "group", "aria-label": langTranslation }, index.h("duet-icon", { name: "navigation-language", margin: "none", color: "secondary" }), index.h("slot", null))));
  }
  get element() { return index.getElement(this); }
};
DuetSlideoutLang.style = duetSlideoutLangCss;

exports.duet_slideout_lang = DuetSlideoutLang;
