// @duetds/components/lib/cjs/keyboard-utils-bb32119b.js
/*!
 * Built with Duet Design System
 */
'use strict';

const isArrowLeftKey = (e) => e.keyCode === 37;
const isArrowUpKey = (e) => e.keyCode === 38;
const isArrowRightKey = (e) => e.keyCode === 39;
const isArrowDownKey = (e) => e.keyCode === 40;
const isArrowKey = (e) => isArrowDownKey(e) || isArrowRightKey(e) || isArrowUpKey(e) || isArrowLeftKey(e);
const isEscapeKey = (e) => e.key === "Escape" || e.key === "Esc" || e.keyCode === 27;
const isTabKey = (e) => e.key === "Tab" || e.keyCode === 9;
const isBackspaceKey = (e) => e.keyCode === 8;
const isSpaceKey = (e) => e.key === "Space" || e.keyCode === 32;
const isPlusKey = (e) => e.key === "+" || e.keyCode === 43;
const isMinusKey = (e) => e.key === "-" || e.keyCode === 45;
const isQuestionKey = (e) => e.key === "?";
const isEnterKey = (e) => e.key === "Enter" || e.keyCode === 13;
const isKeyboardClick = (e) => isEnterKey(e) || isSpaceKey(e);
const isNumber = (e) => (e.keyCode >= 48 && e.keyCode <= 57) || (e.keyCode >= 96 && e.keyCode <= 105);

exports.isArrowDownKey = isArrowDownKey;
exports.isArrowKey = isArrowKey;
exports.isArrowLeftKey = isArrowLeftKey;
exports.isArrowRightKey = isArrowRightKey;
exports.isArrowUpKey = isArrowUpKey;
exports.isBackspaceKey = isBackspaceKey;
exports.isEnterKey = isEnterKey;
exports.isEscapeKey = isEscapeKey;
exports.isKeyboardClick = isKeyboardClick;
exports.isMinusKey = isMinusKey;
exports.isNumber = isNumber;
exports.isPlusKey = isPlusKey;
exports.isQuestionKey = isQuestionKey;
exports.isSpaceKey = isSpaceKey;
exports.isTabKey = isTabKey;
