// @duetds/components/lib/collection/components/duet-list-item/duet-list-item.js
/*!
 * Built with Duet Design System
 */
import { h, Host } from "@stencil/core";
/**
 * @slot label - The list item component’s label content. Use this to put content into description list’s dt element.
 * @slot value - The list item component’s value content. Use this to put content into description list’s dd element.
 */
export class DuetListItem {
  constructor() {
    this.responsive = false;
    this.breakpoint = "small";
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, { class: {
        "duet-no-stacking": !this.responsive,
        "duet-list-item-breakpoint-x-small": this.breakpoint === "x-small",
        "duet-list-item-breakpoint-small": this.breakpoint === "small",
        "duet-list-item-breakpoint-medium": this.breakpoint === "medium",
        "duet-list-item-breakpoint-large": this.breakpoint === "large",
        "duet-list-item-breakpoint-x-large": this.breakpoint === "x-large",
      } }, h("dt", null, h("slot", { name: "label" })), h("dd", null, h("slot", { name: "value" }))));
  }
  static get is() { return "duet-list-item"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["duet-list-item.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["duet-list-item.css"]
    };
  }
  static get properties() {
    return {
      "responsive": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "internal",
              "text": "Enable or disable the automatic responsive behaviour of the list item.\nThis property is used internally by List, and should not be set by the developer."
            }],
          "text": ""
        },
        "attribute": "responsive",
        "reflect": false,
        "defaultValue": "false"
      },
      "breakpoint": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetListBreakpoint",
          "resolved": "\"large\" | \"medium\" | \"small\" | \"x-large\" | \"x-small\"",
          "references": {
            "DuetListBreakpoint": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "internal",
              "text": "Switch the breakpoint used to trigger the content stacking.\nThis property is used internally by List, and should not be set by the developer."
            }],
          "text": ""
        },
        "attribute": "breakpoint",
        "reflect": false,
        "defaultValue": "\"small\""
      }
    };
  }
}
