// @duetds/components/lib/dist-custom-elements/duet-radio-group.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, h, d as Host } from './p-fdc1ca33.js';
import { i as inheritGlobalTheme } from './p-4ebf1618.js';
import { f as findCheckedOption, w as watchForOptions } from './p-12721178.js';
import { d as defineCustomElement$6 } from './p-717d775c.js';
import { d as defineCustomElement$5 } from './p-a15eaed8.js';
import { d as defineCustomElement$4 } from './p-0674039c.js';
import { d as defineCustomElement$3 } from './p-d3213a33.js';
import { d as defineCustomElement$2 } from './p-1344c068.js';

const duetRadioGroupCss = "*.sc-duet-radio-group,*.sc-duet-radio-group::after,*.sc-duet-radio-group::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}.sc-duet-radio-group-h{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:block;width:100%}.horizontal.sc-duet-radio-group-h .duet-radio-group.sc-duet-radio-group{display:block}@media (min-width: 48em){.horizontal.sc-duet-radio-group-h .duet-radio-group.sc-duet-radio-group{display:flex}}.horizontal.sc-duet-radio-group-h .duet-radio-group.duet-no-stacking.sc-duet-radio-group{display:flex}.deut-radio-group.sc-duet-radio-group{width:100%}";

const DuetRadioGroup$1 = /*@__PURE__*/ proxyCustomElement(class extends H {
  constructor() {
    super();
    this.__registerHost();
    /**
     * Component event handling.
     */
    this.onSelect = (ev) => {
      const selectedRadio = ev.target;
      if (selectedRadio) {
        this.value = selectedRadio.value;
      }
      // Prevent events from bubbling up the DOM tree in case
      // this group is a sub-group of another choice group.
      ev.cancelBubble = true;
    };
    this.direction = "vertical";
    this.margin = "auto";
    this.theme = "";
    this.error = "";
    this.tooltip = "";
    this.tooltipDirection = "auto";
    this.name = undefined;
    this.label = "Label";
    this.labelHidden = false;
    this.caption = "";
    this.responsive = false;
    this.value = undefined;
    this.disabled = undefined;
  }
  /**
   * Component lifecycle events.
   */
  async connectedCallback() {
    const el = this.element;
    if (this.value === undefined) {
      const radio = findCheckedOption(el, "duet-radio");
      if (radio !== undefined) {
        await radio.componentOnReady();
        if (this.value === undefined) {
          this.value = radio.value;
        }
      }
    }
    this.mutationO = watchForOptions(el, "duet-radio", newOption => {
      if (newOption !== undefined) {
        newOption.componentOnReady().then(() => {
          this.value = newOption.value;
        });
      }
      else {
        this.updateRadios();
      }
    });
    this.updateRadios();
  }
  disconnectedCallback() {
    if (this.mutationO) {
      this.mutationO.disconnect();
      this.mutationO = undefined;
    }
  }
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * Local methods.
   */
  async updateRadios() {
    /**
     * Make sure we get all radios first so values are up to date prior
     * to caching the radio group value
     */
    const radios = Array.from(this.element.querySelectorAll("duet-radio"));
    const { value } = this;
    let hasChecked = false;
    // Walk the DOM in reverse order, since the last selected one wins
    for (const radio of radios) {
      radio.name = this.name;
      radio.groupDisabled = this.disabled;
      radio.groupDirection = this.direction;
      radio.groupResponsive = this.responsive;
      if (!hasChecked && radio.value === value) {
        // Correct value for this radio, but this radio isn't checked yet
        // and we haven't found a checked yet
        hasChecked = true;
        radio.checked = true;
      }
      else {
        // This radio doesn't have the correct value or the radio group has been already checked
        radio.checked = false;
      }
    }
    // Reset value if
    if (!hasChecked) {
      this.value = undefined;
    }
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, { onDuetChange: this.onSelect, class: { horizontal: this.direction === "horizontal" } }, h("duet-fieldset", { label: this.label, labelHidden: this.labelHidden, caption: this.caption, error: this.error, margin: this.margin, theme: this.theme }, this.tooltip && (h("duet-tooltip", { slot: "tooltip", direction: this.tooltipDirection, accessibleInputLabel: this.label }, this.tooltip)), h("div", { class: { "duet-radio-group": true, "duet-no-stacking": !this.responsive } }, h("slot", null)))));
  }
  get element() { return this; }
  static get watchers() { return {
    "name": ["updateRadios"],
    "value": ["updateRadios"],
    "disabled": ["updateRadios"],
    "direction": ["updateRadios"],
    "responsive": ["updateRadios"]
  }; }
  static get style() { return duetRadioGroupCss; }
}, [6, "duet-radio-group", {
    "direction": [1],
    "margin": [1],
    "theme": [1025],
    "error": [1],
    "tooltip": [1],
    "tooltipDirection": [1, "tooltip-direction"],
    "name": [1],
    "label": [1],
    "labelHidden": [4, "label-hidden"],
    "caption": [1],
    "responsive": [4],
    "value": [1537],
    "disabled": [4]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["duet-radio-group", "duet-caption", "duet-fieldset", "duet-icon", "duet-tooltip", "duet-visually-hidden"];
  components.forEach(tagName => { switch (tagName) {
    case "duet-radio-group":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DuetRadioGroup$1);
      }
      break;
    case "duet-caption":
      if (!customElements.get(tagName)) {
        defineCustomElement$6();
      }
      break;
    case "duet-fieldset":
      if (!customElements.get(tagName)) {
        defineCustomElement$5();
      }
      break;
    case "duet-icon":
      if (!customElements.get(tagName)) {
        defineCustomElement$4();
      }
      break;
    case "duet-tooltip":
      if (!customElements.get(tagName)) {
        defineCustomElement$3();
      }
      break;
    case "duet-visually-hidden":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const DuetRadioGroup = DuetRadioGroup$1;
const defineCustomElement = defineCustomElement$1;

export { DuetRadioGroup, defineCustomElement };
