// @duetds/components/lib/dist-custom-elements/duet-tab.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, h } from './p-fdc1ca33.js';
import { a as getLocaleString } from './p-076452be.js';

const duetTabCss = ".sc-duet-tab-h{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:block;width:100%}.duet-tab.sc-duet-tab{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:none;width:100%}.duet-tab.selected.sc-duet-tab{display:block}";

const DuetTab$1 = /*@__PURE__*/ proxyCustomElement(class extends H {
  constructor() {
    super();
    this.__registerHost();
    this.label = undefined;
    this.caption = undefined;
    this.selected = undefined;
    this.accessibleDescriptionDefault = null;
    this.accessibleDescription = this.accessibleDescriptionDefault
      ? getLocaleString(this.accessibleDescriptionDefault)
      : null;
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    {
      return (h("div", { role: "tabpanel", class: { "duet-tab": true, selected: this.selected }, "aria-description": this.accessibleDescription }, h("slot", null)));
    }
  }
  static get style() { return duetTabCss; }
}, [6, "duet-tab", {
    "label": [513],
    "caption": [513],
    "selected": [516],
    "accessibleDescriptionDefault": [1, "accessible-description-defaults"],
    "accessibleDescription": [1, "accessible-description"]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["duet-tab"];
  components.forEach(tagName => { switch (tagName) {
    case "duet-tab":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DuetTab$1);
      }
      break;
  } });
}

const DuetTab = DuetTab$1;
const defineCustomElement = defineCustomElement$1;

export { DuetTab, defineCustomElement };
