// @duetds/components/lib/dist-custom-elements/duet-toolbar-dropdown-link.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, h, d as Host } from './p-fdc1ca33.js';
import { i as inheritGlobalTheme } from './p-4ebf1618.js';
import { b as isEscapeKey } from './p-d8549efe.js';
import { d as defineCustomElement$2 } from './p-0674039c.js';

const duetToolbarDropdownLinkCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;width:100%;height:100%}.duet-toolbar-dropdown-link{display:flex;flex-direction:row;gap:0.5rem;align-items:center;justify-content:flex-start;height:100%;padding:0.5rem 20px;font-family:\"localtapiola-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";font-size:0.875rem;font-weight:400;line-height:1.25;color:rgb(0, 41, 77);text-align:center;text-decoration:none}.duet-toolbar-dropdown-link.duet-theme-turva{font-family:\"turva-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";color:rgb(23, 28, 58)}.duet-toolbar-dropdown-link:focus{outline:0;box-shadow:0 0 0 4px rgba(0, 119, 179, 0.3), 0 0 0 1px rgb(0, 119, 179)}.duet-toolbar-dropdown-link:focus.duet-theme-turva{box-shadow:0 0 0 4px rgba(23, 28, 58, 0.2), 0 0 0 1px rgb(23, 28, 58)}.duet-toolbar-dropdown-link.active,.duet-toolbar-dropdown-link.active:hover{color:rgb(255, 255, 255);background:rgb(0, 119, 179)}.duet-toolbar-dropdown-link.active.duet-theme-turva,.duet-toolbar-dropdown-link.active:hover.duet-theme-turva{color:rgb(255, 255, 255);background:rgb(198, 12, 48)}.duet-toolbar-dropdown-link:hover{background:rgb(230, 242, 248)}.duet-toolbar-dropdown-link:hover.duet-theme-turva{background:rgb(228, 228, 230)}";

const DuetToolbarDropdownLink$1 = /*@__PURE__*/ proxyCustomElement(class extends H {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.accessibleLabel = undefined;
    this.theme = "";
    this.targetLanguage = undefined;
    this.href = "#";
    this.icon = undefined;
    this.active = false;
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  handleEscape(e) {
    if (isEscapeKey(e)) {
      const dropdown = this.element.closest("duet-toolbar-dropdown");
      dropdown.open = false;
      dropdown.setFocus();
    }
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, null, h("a", { class: {
        "duet-toolbar-dropdown-link": true,
        "duet-theme-turva": this.theme === "turva",
        active: this.active,
      }, href: this.href, lang: this.targetLanguage, hreflang: this.targetLanguage, role: "menuitem", "aria-label": this.accessibleLabel, "aria-current": this.active ? "true" : "false", onKeyUp: e => this.handleEscape(e) }, this.icon && h("duet-icon", { name: this.icon, size: "x-small", margin: "none" }), h("slot", null))));
  }
  get element() { return this; }
  static get style() { return duetToolbarDropdownLinkCss; }
}, [1, "duet-toolbar-dropdown-link", {
    "accessibleLabel": [1, "accessible-label"],
    "theme": [1025],
    "targetLanguage": [1, "target-language"],
    "href": [1025],
    "icon": [1537],
    "active": [1028]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["duet-toolbar-dropdown-link", "duet-icon"];
  components.forEach(tagName => { switch (tagName) {
    case "duet-toolbar-dropdown-link":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DuetToolbarDropdownLink$1);
      }
      break;
    case "duet-icon":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const DuetToolbarDropdownLink = DuetToolbarDropdownLink$1;
const defineCustomElement = defineCustomElement$1;

export { DuetToolbarDropdownLink, defineCustomElement };
