// @duetds/components/lib/dist-custom-elements/duet-nav.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, h, d as Host } from './p-fdc1ca33.js';
import { i as inheritGlobalTheme } from './p-4ebf1618.js';

const duetNavCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;width:100%}slot[name=mobile]{display:inherit}slot[name=desktop]{display:none}@media (min-width: 62em){slot[name=mobile]{display:none}slot[name=desktop]{display:inherit}}";

const DuetNav$1 = /*@__PURE__*/ proxyCustomElement(class extends H {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.theme = "";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, null, h("div", null, h("slot", null), h("slot", { name: "mobile" }), h("slot", { name: "desktop" }))));
  }
  get element() { return this; }
  static get style() { return duetNavCss; }
}, [1, "duet-nav", {
    "theme": [1025]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["duet-nav"];
  components.forEach(tagName => { switch (tagName) {
    case "duet-nav":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DuetNav$1);
      }
      break;
  } });
}

const DuetNav = DuetNav$1;
const defineCustomElement = defineCustomElement$1;

export { DuetNav, defineCustomElement };
