// @duetds/components/lib/dist-custom-elements/duet-promo-card.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, h, d as Host } from './p-fdc1ca33.js';
import { i as inheritGlobalTheme } from './p-4ebf1618.js';
import { g as getColorByName } from './p-de68431d.js';
import { d as defineCustomElement$3 } from './p-628b8684.js';
import { d as defineCustomElement$2 } from './p-75c1e787.js';

const duetPromoCardCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;margin-right:4px !important;margin-bottom:2px !important;display:block;width:auto}:host:last-child,:host:last-of-type{margin-right:0 !important}:host(.duet-m-0){margin:0 !important}.duet-promo-card{display:flex;flex-direction:column;width:100%;height:100%;font-family:\"localtapiola-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";font-style:normal;color:rgb(0, 41, 77);text-align:left;border-radius:20px}.duet-promo-card.duet-p-0{padding:0 !important}.duet-promo-card.duet-m-0{margin:0 !important}.duet-promo-card .image{position:relative;width:100%;overflow:hidden;border-top-left-radius:20px;border-top-right-radius:20px;object-fit:cover}.duet-promo-card .content{display:flex;flex-direction:column;flex-grow:1;gap:8px;padding:20px}@media (min-width: 36em){.duet-promo-card .content{padding:28px}}.duet-promo-card .action{display:flex;flex-grow:1;align-items:flex-end;padding-top:16px}.duet-promo-card.no-bg-color .content{padding:20px 0 0}.duet-promo-card.no-bg-color .image{border-radius:16px}.duet-promo-card.duet-theme-turva{font-family:\"turva-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\"}";

const DuetPromoCard$1 = /*@__PURE__*/ proxyCustomElement(class extends H {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.theme = "";
    this.margin = "auto";
    this.backgroundColor = "";
    this.textColor = "secondary";
    this.image = "";
    this.headingLevel = "h3";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
    this.hasPreHeadingSlot = !!this.element.querySelector('[slot="pre-heading"]');
    this.hasHeadingSlot = !!this.element.querySelector('[slot="heading"]');
    this.hasDescriptionSlot = !!this.element.querySelector('[slot="description"]');
    this.hasActionSlot = !!this.element.querySelector('[slot="action"]');
  }
  /**
   * render() function
   * Always the last one of the class.
   */
  render() {
    const backgroundColor = getColorByName(this.backgroundColor, this.theme);
    const bannerStyles = {
      backgroundColor: backgroundColor ? backgroundColor : "none",
    };
    return (h(Host, { class: { "duet-m-0": this.margin === "none" } }, h("aside", { class: {
        "duet-promo-card": true,
        "duet-theme-turva": this.theme === "turva",
        "no-bg-color": !this.backgroundColor,
      }, style: bannerStyles }, h("img", { class: "image", src: this.image, alt: "", role: "presentation" }), h("div", { class: "content" }, this.hasPreHeadingSlot && (h("div", { class: "pre-heading" }, h("slot", { name: "pre-heading" }))), this.hasHeadingSlot && (h("div", { class: "heading" }, h("duet-heading", { theme: this.theme, level: this.headingLevel, "visual-level": "h3", margin: "none", color: this.textColor }, h("slot", { name: "heading" })))), this.hasDescriptionSlot && (h("div", { class: "description" }, h("duet-paragraph", { theme: this.theme, margin: "none", color: this.textColor }, h("slot", { name: "description" })))), this.hasActionSlot && (h("div", { class: "action" }, h("slot", { name: "action" })))))));
  }
  get element() { return this; }
  static get style() { return duetPromoCardCss; }
}, [1, "duet-promo-card", {
    "theme": [1025],
    "margin": [1],
    "backgroundColor": [1, "background-color"],
    "textColor": [1, "text-color"],
    "image": [1],
    "headingLevel": [1, "heading-level"]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["duet-promo-card", "duet-heading", "duet-paragraph"];
  components.forEach(tagName => { switch (tagName) {
    case "duet-promo-card":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DuetPromoCard$1);
      }
      break;
    case "duet-heading":
      if (!customElements.get(tagName)) {
        defineCustomElement$3();
      }
      break;
    case "duet-paragraph":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const DuetPromoCard = DuetPromoCard$1;
const defineCustomElement = defineCustomElement$1;

export { DuetPromoCard, defineCustomElement };
