// @duetds/components/lib/dist-custom-elements/duet-slideout-lang.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, h, d as Host } from './p-fdc1ca33.js';
import { i as inheritGlobalTheme } from './p-4ebf1618.js';
import { g as getLanguage, c as connectLanguageChangeObserver, d as disconnectLanguageChangeObserver, a as getLocaleString } from './p-076452be.js';
import { d as defineCustomElement$2 } from './p-0674039c.js';

const duetSlideoutLangCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}.duet-slideout-lang{display:flex;gap:20px;align-items:center;justify-content:flex-start;width:min-content}::slotted(*){position:relative}::slotted(:not(:first-child)){padding-right:0;padding-left:0.5rem}::slotted(:not(:last-child))::after{position:absolute;top:15%;right:-16px;font-family:\"localtapiola-sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, Helvetica, Arial, sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\";font-size:16px;font-weight:600;line-height:150%;color:rgb(0, 41, 77);content:\"|\"}.duet-theme-turva ::slotted(:not(:last-child))::after{color:rgb(23, 28, 58)}";

const DuetSlideoutLang$1 = /*@__PURE__*/ proxyCustomElement(class extends H {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    /**
     * Own Properties.
     */
    this.chooseLangObject = {
      en: "Choose language",
      fi: "Valitse kieli",
      sv: "Välj språk",
    };
    this.language = getLanguage();
    this.theme = "";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  connectedCallback() {
    connectLanguageChangeObserver(this);
  }
  disconnectedCallback() {
    disconnectLanguageChangeObserver(this);
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    const langTranslation = getLocaleString(this.chooseLangObject);
    return (h(Host, null, h("div", { class: "duet-slideout-lang", role: "group", "aria-label": langTranslation }, h("duet-icon", { name: "navigation-language", margin: "none", color: "secondary" }), h("slot", null))));
  }
  get element() { return this; }
  static get style() { return duetSlideoutLangCss; }
}, [1, "duet-slideout-lang", {
    "theme": [1025],
    "language": [32]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["duet-slideout-lang", "duet-icon"];
  components.forEach(tagName => { switch (tagName) {
    case "duet-slideout-lang":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DuetSlideoutLang$1);
      }
      break;
    case "duet-icon":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const DuetSlideoutLang = DuetSlideoutLang$1;
const defineCustomElement = defineCustomElement$1;

export { DuetSlideoutLang, defineCustomElement };
