// @duetds/components/lib/dist-custom-elements/p-4fcae101.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, h, d as Host } from './p-fdc1ca33.js';
import { d as defineCustomElement$1 } from './p-1344c068.js';

const DuetUploadAriaStatus = /*@__PURE__*/ proxyCustomElement(class extends H {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    /**
     * Private functions
     */
    this.replacePlaceHolders = (statusString) => {
      const addFiles = (amount) => {
        return amount > 1 ? `${amount} ${this.statusMessageLabel.files}` : `${amount} ${this.statusMessageLabel.file}`;
      };
      return statusString
        .replace("{filesUploaded}", addFiles(this.valid))
        .replace("{filesInProgress}", addFiles(this.inprogress))
        .replace("{filesTotal}", addFiles(this.total))
        .replace("{filesWithErrors}", addFiles(this.invalid));
    };
    this.getStatusMessage = () => {
      const { done, inProgress, doneWithErrors, inProgressWithErrors } = this.statusMessageLabel;
      const hasErrors = this.invalid > 0;
      const isInProgress = this.inprogress > 0;
      const isInProgressWithErrors = this.inprogress > 0 && hasErrors;
      const isDone = this.valid === this.total && !isInProgress && !hasErrors;
      const isDoneWithErrors = this.valid + this.invalid === this.total && !isInProgress;
      if (isDone) {
        this.statusMessage = this.replacePlaceHolders(done);
        return;
      }
      else if (isInProgressWithErrors) {
        this.statusMessage = this.replacePlaceHolders(inProgressWithErrors);
        return;
      }
      else if (isInProgress) {
        this.statusMessage = this.replacePlaceHolders(inProgress);
        return;
      }
      else if (isDoneWithErrors) {
        this.statusMessage = this.replacePlaceHolders(doneWithErrors);
        return;
      }
      return;
    };
    this.statusMessage = null;
    this.statusMessageLabel = null;
    this.accessibleAriaLive = "polite";
    this.valid = undefined;
    this.total = undefined;
    this.invalid = undefined;
    this.inprogress = undefined;
  }
  watchHandler(newValue, oldValue) {
    if (newValue !== oldValue) {
      this.getStatusMessage();
    }
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, { role: "status", "aria-live": "assertive", "aria-atomic": "true", "aria-relevant": "all" }, h("duet-visually-hidden", null, this.statusMessage)));
  }
  static get watchers() { return {
    "valid": ["watchHandler"],
    "invalid": ["watchHandler"],
    "inprogress": ["watchHandler"],
    "total": ["watchHandler"]
  }; }
}, [1, "duet-upload-aria-status", {
    "statusMessageLabel": [16],
    "accessibleAriaLive": [1, "accessible-aria-live"],
    "valid": [2],
    "total": [2],
    "invalid": [2],
    "inprogress": [2],
    "statusMessage": [32]
  }]);
function defineCustomElement() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["duet-upload-aria-status", "duet-visually-hidden"];
  components.forEach(tagName => { switch (tagName) {
    case "duet-upload-aria-status":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DuetUploadAriaStatus);
      }
      break;
    case "duet-visually-hidden":
      if (!customElements.get(tagName)) {
        defineCustomElement$1();
      }
      break;
  } });
}

export { DuetUploadAriaStatus as D, defineCustomElement as d };
