// @duetds/components/lib/dist-custom-elements/p-bcf1e39a.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, h, d as Host } from './p-fdc1ca33.js';
import { i as inheritGlobalTheme } from './p-4ebf1618.js';
import { c as createID } from './p-c1325e35.js';
import { a as getLocaleString } from './p-076452be.js';
import { d as defineCustomElement$1 } from './p-1344c068.js';

const duetProgressCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;width:100%;height:inherit}.duet-progress{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:block;width:100%;min-height:1px;}.duet-progress .progress-bar{position:relative;display:block;width:100%;height:inherit;min-height:1px;background-color:rgb(243, 249, 252);border-radius:0;box-shadow:none}.duet-theme-turva .duet-progress .progress-bar{background-color:rgb(252, 243, 244)}.duet-progress .progress-bar>span{display:block;height:inherit;text-indent:-9999px;background-color:rgb(0, 119, 179);border-radius:0}.duet-theme-turva .duet-progress .progress-bar>span{background-color:rgb(198, 12, 48)}.duet-progress progress[value]{display:flex;width:100%;height:inherit;min-height:1px;color:rgb(0, 119, 179);border:0 none;-webkit-appearance:none;-moz-appearance:none;appearance:none}.duet-theme-turva .duet-progress progress[value]{color:rgb(198, 12, 48)}.duet-progress progress[value]::-webkit-progress-bar{background-color:rgb(243, 249, 252);border-radius:0}.duet-theme-turva .duet-progress progress[value]::-webkit-progress-bar{background-color:rgb(252, 243, 244)}.duet-progress progress[value]::-webkit-progress-value{background-color:rgb(0, 119, 179);background-size:35px 20px, 100% 100%, 100% 100%;border-radius:0;-webkit-animation:animate-stripes 5s linear infinite;animation:animate-stripes 5s linear infinite}.duet-theme-turva .duet-progress progress[value]::-webkit-progress-value{background-color:rgb(198, 12, 48)}.duet-progress progress[value] ::-moz-progress-bar{background-color:rgb(243, 249, 252)}.duet-theme-turva .duet-progress progress[value] ::-moz-progress-bar{background-color:rgb(252, 243, 244)}";

const DuetProgress = /*@__PURE__*/ proxyCustomElement(class extends H {
  constructor() {
    super();
    this.__registerHost();
    /**
     * Own Properties
     */
    this.progressId = createID("DuetProgress");
    this.progressLabelId = createID("DuetProgressLabel");
    this.accessibleDescribedBy = undefined;
    this.accessibleDetails = undefined;
    this.accessibleLabelledBy = undefined;
    this.accessibleDescription = undefined;
    this.accessibleLabel = undefined;
    this.accessibleLabelUploadDefaults = {
      fi: ["lähetys valmis", "lähetys käynnissä"],
      sv: ["Uppladdningen är klar", "Uppladdning pågår"],
      en: ["upload completed", "upload in progress"],
    };
    this.accessibleLabelUpload = getLocaleString(this.accessibleLabelUploadDefaults);
    this.progress = 0;
    this.caption = undefined;
    this.height = "10px";
    this.theme = "";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, null, h("div", { "aria-live": this.accessibleLabel ? "polite" : "off", "aria-hidden": this.accessibleLabel ? "false" : "true", "aria-atomic": "true", style: {
        height: this.height,
      }, id: this.progressId, class: {
        "duet-progress": true,
        "duet-progress-done": this.progress === 100,
        "duet-progress-start": this.progress === 0,
        "duet-progress-inprogress": this.progress < 99 && this.progress !== 0,
        "duet-theme-turva": this.theme === "turva",
      } }, h("duet-visually-hidden", null, h("label", { htmlFor: this.progressLabelId }, this.progress !== 0 && (h("span", null, this.accessibleLabel, ", ", this.accessibleLabelUpload[this.progress < 99 ? 1 : 0])))), this.caption && (h("div", { class: "duet-progress-caption" }, h("span", null, this.caption))), h("progress", { id: this.progressLabelId, "aria-describedby": this.accessibleDescribedBy, "aria-labelledby": this.accessibleLabelledBy, "aria-details": this.accessibleDetails, "aria-description": this.accessibleDescription, role: "progressbar", value: this.progress, max: "100", "aria-valuemin": "0", "aria-valuemax": "100" }, h("div", { class: "progress-bar" }, h("span", { style: { width: `${this.progress}%;` } }, "Progress: ", this.progress, "%"))))));
  }
  get element() { return this; }
  static get style() { return duetProgressCss; }
}, [0, "duet-progress", {
    "accessibleDescribedBy": [1, "accessible-described-by"],
    "accessibleDetails": [1, "accessible-details"],
    "accessibleLabelledBy": [1, "accessible-labelled-by"],
    "accessibleDescription": [1, "accessible-description"],
    "accessibleLabel": [1, "accessible-label"],
    "accessibleLabelUploadDefaults": [1, "accessible-label-upload-defaults"],
    "accessibleLabelUpload": [1, "accessible-label-upload"],
    "progress": [1538],
    "caption": [1],
    "height": [1],
    "theme": [1025]
  }]);
function defineCustomElement() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["duet-progress", "duet-visually-hidden"];
  components.forEach(tagName => { switch (tagName) {
    case "duet-progress":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DuetProgress);
      }
      break;
    case "duet-visually-hidden":
      if (!customElements.get(tagName)) {
        defineCustomElement$1();
      }
      break;
  } });
}

export { DuetProgress as D, defineCustomElement as d };
