// @duetds/components/lib/dist-custom-elements/p-d7b4d591.js
/*!
 * Built with Duet Design System
 */
import { p as proxyCustomElement, H, h } from './p-fdc1ca33.js';
import { i as inheritGlobalTheme } from './p-4ebf1618.js';
import { g as getColorByName } from './p-de68431d.js';

const duetDividerCss = "*,*::after,*::before{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none}:host{box-sizing:border-box;padding:0;margin:0;background:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;appearance:none;display:block;width:100%}.duet-divider{-webkit-user-select:none;user-select:none;display:block;width:100%;height:1px;margin-top:36px !important;margin-bottom:28px;background:rgb(225, 227, 230)}.duet-divider.duet-p-0{padding:0 !important}.duet-divider.duet-m-0{margin:0 !important}.duet-divider.duet-m-s{margin-bottom:16px !important;margin-top:0 !important}.duet-divider.duet-theme-turva{background:rgb(228, 228, 230)}";

const DuetDivider = /*@__PURE__*/ proxyCustomElement(class extends H {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.theme = "";
    this.margin = "auto";
    this.color = "";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    const background = getColorByName(this.color);
    return (h("div", { class: {
        "duet-m-0": this.margin === "none",
        "duet-m-s": this.margin === "small",
        "duet-divider": true,
        "duet-theme-turva": this.theme === "turva",
      }, style: { background } }));
  }
  get element() { return this; }
  static get style() { return duetDividerCss; }
}, [1, "duet-divider", {
    "theme": [1025],
    "margin": [1],
    "color": [1]
  }]);
function defineCustomElement() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["duet-divider"];
  components.forEach(tagName => { switch (tagName) {
    case "duet-divider":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, DuetDivider);
      }
      break;
  } });
}

export { DuetDivider as D, defineCustomElement as d };
