// @duetds/components/lib/dist-custom-elements/p-de68431d.js
/*!
 * Built with Duet Design System
 */
import { m as mediaQueryXxSmall, a as mediaQueryXSmall, b as mediaQuerySmall, c as mediaQueryMedium, d as mediaQueryLarge, e as mediaQueryXLarge, f as mediaQueryXxLarge, g as mediaQueryXxxLarge, h as mediaQueryXxxSmall, t as tokens } from './p-e8475b3c.js';

const allowedThemes = ["turva"];
function isDashCase(str) {
  const dash = str.includes("-");
  const barredString = str.includes(" ") || str.includes("_");
  const barredStartingString = str.charAt(0) === "-";
  return dash && !barredString && !barredStartingString;
}
function capitalize(str = "") {
  return `${str.charAt(0).toUpperCase()}${str.slice(1)}`;
}
function lowerCaseFirst(str = "") {
  return `${str.charAt(0).toLowerCase()}${str.slice(1)}`;
}
function camelize(str = "") {
  const arr = str.split("-");
  //loop through each element of the array and capitalize the first letter.'
  arr.forEach((string, index) => {
    arr[index] = capitalize(string);
  });
  return arr.join("");
}
function getTheme(theme = "") {
  if (allowedThemes.includes(theme.toLowerCase())) {
    return theme.toLowerCase();
  }
  return "";
}
function capitalizeFirstLetter(string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}
function hasType(t, str = "") {
  return str.indexOf(capitalizeFirstLetter(t)) === 0 || str.indexOf(t) === 0;
}
function addPrefix(prefix, str = "") {
  if (str && !isDashCase(str) && !hasType(prefix, str)) {
    return `${prefix}${capitalize(str)}`;
  }
  return str;
}
function addTheme(str, theme) {
  const hasTheme = getTheme(theme).length;
  if (hasTheme && str.indexOf(theme) === -1) {
    return `${str}${capitalize(theme)}`;
  }
  return str;
}
function getColorByName(name, theme = "") {
  if (!name) {
    return undefined;
  }
  const defaultColor = convertToColorName(name);
  const themedColor = convertToColorName(name, theme);
  // lets just try to get the color first to save some cycles
  let resolvedToken = tokens[themedColor];
  if (!resolvedToken) {
    resolvedToken = tokens[defaultColor];
  }
  return resolvedToken;
}
function convertToColorName(name = "", theme = "") {
  let colorResolution = name;
  const dash = isDashCase(name);
  if (dash) {
    colorResolution = camelize(colorResolution);
  }
  colorResolution = addPrefix("color", colorResolution);
  colorResolution = addTheme(colorResolution, theme);
  // takes care of the case where colorPrimaryDark ends up with ColorPrimaryDark
  colorResolution = lowerCaseFirst(colorResolution);
  return colorResolution;
}
function rgbaColorFromRgb(color, opacity) {
  if (!color || (opacity !== 0 && !opacity)) {
    return undefined;
  }
  const f = color.split(",");
  const R = parseInt(f[0].slice(4)), G = parseInt(f[1]), B = parseInt(f[2]);
  return `rgba(${R},${G},${B},${opacity})`;
}
function shadeRGBColor(color, percent) {
  const f = color.split(","), t = percent < 0 ? 0 : 255, p = percent < 0 ? percent * -1 : percent;
  let R = parseInt(f[0].slice(4)), G = parseInt(f[1]), B = parseInt(f[2]);
  R = Math.round((t - R) * p) + R;
  G = Math.round((t - G) * p) + G;
  B = Math.round((t - B) * p) + B;
  R = R < 0 ? 0 : R;
  R = R > 255 ? 255 : R;
  G = G < 0 ? 0 : G;
  G = G > 255 ? 255 : G;
  B = B < 0 ? 0 : B;
  B = B > 255 ? 255 : B;
  return `rgb(${R},${G},${B})`;
}
// remove wrapping speech marks.
// media query tokens are wrapped in speech marks,
// which need to be removed before use with matchMedia.
const unwrap = (str) => str.substring(1, str.length - 1);
const breakpointToToken = {
  "xxx-small": unwrap(mediaQueryXxxSmall),
  "xx-small": unwrap(mediaQueryXxSmall),
  "x-small": unwrap(mediaQueryXSmall),
  small: unwrap(mediaQuerySmall),
  medium: unwrap(mediaQueryMedium),
  large: unwrap(mediaQueryLarge),
  "x-large": unwrap(mediaQueryXLarge),
  "xx-large": unwrap(mediaQueryXxLarge),
  "xxx-large": unwrap(mediaQueryXxxLarge),
};
const breakpointTokensUpTo = Object.keys(breakpointToToken).filter(key => key.indexOf("x-small") !== -1);
Object.keys(breakpointToToken).filter(key => !breakpointTokensUpTo.includes(key));

export { breakpointToToken as a, breakpointTokensUpTo as b, capitalizeFirstLetter as c, getColorByName as g, rgbaColorFromRgb as r, shadeRGBColor as s };
