// @duetds/components/lib/collection/components/duet-badge/duet-badge.js
/*!
 * Built with Duet Design System
 */
import { h, Host } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
export class DuetBadge {
  constructor() {
    this.theme = "";
    this.variation = "default";
    this.background = "normal";
    this.margin = "auto";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function
   * Always the last one of the class.
   */
  render() {
    const backgroundClass = `background-${this.background}`;
    return (h(Host, { class: { "duet-m-0": this.margin === "none" } }, h("div", { class: {
        "duet-badge": true,
        [this.variation]: true,
        [backgroundClass]: true,
        "duet-theme-turva": this.theme === "turva",
      } }, h("slot", null))));
  }
  static get is() { return "duet-badge"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["duet-badge.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["duet-badge.css"]
    };
  }
  static get properties() {
    return {
      "theme": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetTheme",
          "resolved": "\"\" | \"default\" | \"turva\"",
          "references": {
            "DuetTheme": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Theme of the badge."
        },
        "attribute": "theme",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "variation": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetStatus",
          "resolved": "\"danger\" | \"default\" | \"success\" | \"warning\"",
          "references": {
            "DuetStatus": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Style variation of the badge."
        },
        "attribute": "variation",
        "reflect": false,
        "defaultValue": "\"default\""
      },
      "background": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "\"normal\" | \"strong\"",
          "resolved": "\"normal\" | \"strong\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "default",
              "text": "\"normal\""
            }],
          "text": "Variation of the background. \"strong\" results in a more prominent background color."
        },
        "attribute": "background",
        "reflect": false,
        "defaultValue": "\"normal\""
      },
      "margin": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetMargin",
          "resolved": "\"auto\" | \"none\"",
          "references": {
            "DuetMargin": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Controls the margin of the component."
        },
        "attribute": "margin",
        "reflect": false,
        "defaultValue": "\"auto\""
      }
    };
  }
}
