// @duetds/components/lib/collection/components/duet-caption/duet-caption.js
/*!
 * Built with Duet Design System
 */
import { h, Host } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
export class DuetCaption {
  constructor() {
    this.theme = "";
    this.margin = "auto";
    this.selected = false;
    this.size = "medium";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    return (h(Host, { class: {
        "duet-caption": true,
        "duet-caption-selected": this.selected,
        "duet-caption-small": this.size === "small",
        "duet-theme-turva": this.theme === "turva",
        "duet-m-0": this.margin === "none",
      } }, h("slot", null)));
  }
  static get is() { return "duet-caption"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["duet-caption.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["duet-caption.css"]
    };
  }
  static get properties() {
    return {
      "theme": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetTheme",
          "resolved": "\"\" | \"default\" | \"turva\"",
          "references": {
            "DuetTheme": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Theme of the caption."
        },
        "attribute": "theme",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "margin": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetMargin",
          "resolved": "\"auto\" | \"none\"",
          "references": {
            "DuetMargin": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Controls the margin of the component."
        },
        "attribute": "margin",
        "reflect": false,
        "defaultValue": "\"auto\""
      },
      "selected": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Selected state of the caption. Used inside Choice component."
        },
        "attribute": "selected",
        "reflect": false,
        "defaultValue": "false"
      },
      "size": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetCaptionSize",
          "resolved": "\"medium\" | \"small\"",
          "references": {
            "DuetCaptionSize": {
              "location": "local",
              "path": "/Users/laaksju3/duetds/packages/components/src/components/duet-caption/duet-caption.tsx"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Controls the size of the caption."
        },
        "attribute": "size",
        "reflect": false,
        "defaultValue": "\"medium\""
      }
    };
  }
}
