// @duetds/components/lib/collection/components/duet-logo/duet-logo.js
/*!
 * Built with Duet Design System
 */
import { colorBrandGrayTurva, colorGrayLightest, colorPrimary, colorPrimaryTurva } from "@duetds/tokens";
import { h, Host } from "@stencil/core";
import { inheritGlobalTheme } from "../../common/themeable-component";
import { createID } from "../../utils/create-id";
import { connectLanguageChangeObserver, disconnectLanguageChangeObserver, getLanguage, getLocaleString, } from "../../utils/language-utils";
const TurvaLogo = ({ inverse, logoId }) => {
  const logoColor = inverse ? colorGrayLightest : colorBrandGrayTurva;
  const textColor = inverse ? colorGrayLightest : colorPrimaryTurva;
  return (h("svg", { role: "img", xmlns: "http://www.w3.org/2000/svg", "aria-labelledby": logoId, viewBox: "0 0 305 111" }, h("title", { id: logoId }, "Turva"), h("path", { fill: logoColor, d: "M79 90.1A40.5 40.5 0 0 0 98.3 58H79v32.2z" }), h("path", { fill: logoColor, d: "M45.1 57.9V46.4h52.3a40.5 40.5 0 1 0-30 48.5V58H45z" }), h("path", { fill: textColor, d: "M278.9 80.3v-9.8l-7.4.4c-2.4.3-4.3.6-5.7 1-1.7.4-4 1.3-5 2.3-1 1-2.3 2.8-2.3 4.6 0 2 .8 3.8 2 4.8s3.3 2 5.7 2a22 22 0 0 0 12.7-5.3zm4.2 8H279v-4.1l-3 1.9-2.8 1.5a17.4 17.4 0 0 1-7.4 1.6c-3 0-6.3-.7-8-2.3a9.6 9.6 0 0 1-3.4-7.2 9 9 0 0 1 1.8-5.9c1.1-1.6 3.1-3.3 5.2-4.2 2.1-1 4.7-1.6 7.7-1.9 3.4-.4 6.8-.6 9.8-.7v-1.1a6 6 0 0 0-.7-3.3 8 8 0 0 0-2.6-2.5c-.8-.5-2.2-1-3.4-1.1a25.7 25.7 0 0 0-8.8.4c-1.9.4-3.3 1-5.4 1.7h-.3v-4.3c1.2-.3 2.3-.7 4.5-1a37.7 37.7 0 0 1 12.8 0c1.8.4 2.9 1 4.2 2 1.3 1 2.3 2.3 3 3.8.7 1.5 1 3.2 1 4.7v22zM250.1 56l-14 32.4h-4.8L217.4 56h4.5l12 27.7L245.6 56zM137.7 88.2l-2.7.7c-.9.2-2 .4-4.2.4a9 9 0 0 1-7.2-3c-1.9-1.8-2.6-5-2.6-9.3V60.2h-5.2V56h5.2V47h4.2v9h12.6v4.2h-12.6v14.2c0 1.8 0 4.3.2 5.4 0 1 .3 2.5.8 3.4.5.8 1.4 1.4 2.2 1.8.8.4 2.7.5 4.1.5 1-.1 1.7-.2 2.7-.5l2.2-.6h.3v3.8zM176.4 88.4h-4.2v-3.9c-2 1.6-4.5 2.8-6.3 3.6-1.8.8-3.8 1.3-6 1.3-3.6 0-6-1.2-8-3.4s-2.5-5.5-2.5-9.9v-20h4.2V74c0 1.8.1 3.8.3 5 .2 1.3.5 3 1 3.8a5 5 0 0 0 2.2 2c1 .5 2.2.7 4 .7 1.5 0 3.1-.4 5-1.2 1.7-.8 4.5-2.4 6-3.6V56.1h4.3v32.3zM210.4 61h-.3a21.6 21.6 0 0 0-5.4-.7c-1.9 0-3.6.4-5.3 1.2a31 31 0 0 0-5.4 3.7v23.2h-4.3V56h4.3v5.2c2.3-1.9 5-3.2 6.7-4 1.8-.8 3.6-1.2 5.5-1.2a22.3 22.3 0 0 1 4.2.4v4.4z" })));
};
const i18n = {
  fi: {
    viewBox: "0 0 427 111",
    title: "LähiTapiola",
    path: "M131.6 69.6v1.3l-.3 1c0 .2-.2.4-.3.5l-.5.1h-16c-.5 0-1-.1-1.4-.5-.5-.3-.7-.9-.7-1.7V38.7l.2-.5c.1-.2.3-.3.6-.4l1-.2a15.7 15.7 0 0 1 4.6.2c.3.1.5.2.6.4l.2.5v28h11l.4.1.3.5.2 1 .1 1.3M164.3 70c.2.5.4 1 .4 1.4 0 .4 0 .7-.2.9-.2.2-.5.3-1 .3a25.2 25.2 0 0 1-5.4 0l-.6-.4-.3-.6-2.3-7h-13l-2.3 6.8c0 .3-.2.5-.3.6-.1.2-.3.4-.6.4l-1 .3a25.9 25.9 0 0 1-3.9 0l-1-.5c-.1-.2-.2-.5-.1-.8 0-.4.2-1 .4-1.5l10.7-31 .4-.7c.1-.2.4-.3.7-.4l1.3-.2a47.7 47.7 0 0 1 6.2.2c.4 0 .6.2.8.4l.4.8 10.7 31zm-17.9-38.8c0 1.3-.2 2.2-.7 2.7-.5.5-1.4.7-2.8.7s-2.3-.2-2.8-.7c-.5-.5-.7-1.4-.7-2.6 0-1.2.2-2.1.7-2.6.5-.5 1.4-.8 2.8-.8s2.4.3 2.8.7c.5.5.7 1.4.7 2.6zm2 13.1h-.1l-5 14.9h10l-5-14.9zm9.8-13c0 1.2-.2 2-.7 2.6-.5.5-1.5.7-2.9.7s-2.3-.2-2.8-.7c-.4-.5-.7-1.4-.7-2.6 0-1.2.3-2.1.8-2.6s1.4-.8 2.8-.8 2.3.3 2.8.7c.5.5.7 1.4.7 2.6zM196.4 71.6c0 .2 0 .3-.2.5l-.5.3-1.1.2a14 14 0 0 1-4.5-.2c-.3 0-.5-.2-.6-.3a.7.7 0 0 1-.2-.5V57.5h-13v14c0 .3 0 .4-.2.6l-.5.3-1.1.2a14.3 14.3 0 0 1-4.5-.2c-.3 0-.5-.2-.6-.3a.8.8 0 0 1-.2-.5v-33c0-.1 0-.3.2-.4.1-.2.3-.3.6-.4l1-.2a15.6 15.6 0 0 1 4.6.2l.5.4.2.5v12.8h13V38.7c0-.2 0-.4.2-.5.1-.2.3-.3.6-.4l1-.2a15.9 15.9 0 0 1 4.6.2l.5.4c.2.1.2.3.2.5v32.9zM211.1 71.6c0 .2 0 .3-.2.5l-.5.3-1.1.2a14.3 14.3 0 0 1-4.5-.2c-.3 0-.5-.2-.6-.3a.8.8 0 0 1-.2-.5v-33c0-.1 0-.3.2-.4.1-.2.3-.3.6-.4l1-.2a15.3 15.3 0 0 1 4.6.2l.5.4c.2.1.2.3.2.5v32.9zM242 40.6l-.1 1.3-.2 1-.4.4-.4.2h-8.7v28c0 .3 0 .4-.2.6l-.6.3a14 14 0 0 1-5.6 0c-.3 0-.4-.2-.6-.3a.8.8 0 0 1-.1-.5V43.5h-8.8c-.1 0-.3 0-.4-.2-.1 0-.3-.3-.3-.5l-.2-.9a10.9 10.9 0 0 1 .2-3.6c0-.2.2-.4.3-.5l.4-.1H241l.4.1.4.5.2 1v1.3M270.3 70l.4 1.4c0 .4 0 .7-.2.9-.2.2-.6.3-1 .3a24.3 24.3 0 0 1-5.5 0l-.5-.4-.3-.6-2.4-7h-13l-2.3 6.8-.2.6c-.2.2-.3.4-.6.4a25.9 25.9 0 0 1-5 .2c-.4 0-.7-.2-.9-.4-.2-.2-.2-.5-.2-.8 0-.4.2-1 .4-1.5l10.8-31c0-.3.2-.5.3-.7l.7-.4 1.3-.2a47.8 47.8 0 0 1 6.2.2c.4 0 .7.2.8.4l.4.8 10.8 31zm-16-25.7l-5 14.9h9.9l-5-14.9zM299 48.3c0 2-.3 3.7-.9 5.2a10 10 0 0 1-2.7 3.8c-1.1 1-2.6 1.8-4.3 2.3a20 20 0 0 1-6 .8h-2.9v11.2l-.1.5-.6.3-1.1.2a14 14 0 0 1-4.5-.2c-.3 0-.5-.2-.6-.3a.8.8 0 0 1-.2-.5V40.2c0-.8.3-1.5.7-1.9.4-.4 1-.6 1.7-.6h8.3a28 28 0 0 1 5 .5c1 .2 2.1.6 3.2 1.1a8.8 8.8 0 0 1 4.5 5.1c.3 1.2.5 2.5.5 3.9m-7.4.5c0-1.2-.2-2.2-.6-3-.5-.8-1-1.4-1.6-1.8-.7-.4-1.3-.6-2-.7l-2.2-.1h-3V55h3.2c1.1 0 2-.2 2.9-.5a4.9 4.9 0 0 0 3-3.2c.2-.8.3-1.6.3-2.5M311.6 71.6c0 .2 0 .3-.2.5l-.5.3-1.1.2a14.2 14.2 0 0 1-4.5-.2c-.3 0-.5-.2-.6-.3a.8.8 0 0 1-.2-.5v-33c0-.1 0-.3.2-.4.1-.2.3-.3.6-.4l1-.2a15.2 15.2 0 0 1 4.6.2l.5.4c.2.1.2.3.2.5v32.9zM350.6 54.7c0 2.9-.4 5.5-1.1 7.8-.7 2.2-1.8 4.2-3.2 5.8a13.7 13.7 0 0 1-5.3 3.6 21.3 21.3 0 0 1-14.4.2 12.3 12.3 0 0 1-8-8.8 28 28 0 0 1 0-15.6c.8-2.2 1.8-4.2 3.3-5.7a14 14 0 0 1 5.2-3.7c2.1-.8 4.5-1.2 7.3-1.2 2.7 0 5 .3 7 1s3.7 1.8 5 3.2a14 14 0 0 1 3.1 5.6c.7 2.2 1 4.8 1 7.8m-7.4.4c0-1.8-.1-3.5-.4-5-.3-1.5-.8-2.8-1.5-3.9a7 7 0 0 0-2.7-2.4 9.4 9.4 0 0 0-4.3-1c-1.8 0-3.2.4-4.4 1a8 8 0 0 0-2.8 2.7 11 11 0 0 0-1.5 3.8 27.8 27.8 0 0 0 0 9.8 11 11 0 0 0 1.4 3.9c.7 1 1.6 1.9 2.7 2.5 1.2.5 2.6.8 4.4.8 1.7 0 3.2-.3 4.3-1a7.8 7.8 0 0 0 2.9-2.6c.7-1.1 1.2-2.4 1.5-3.9.3-1.5.4-3 .4-4.7M375.7 69.6a10.5 10.5 0 0 1-.3 2.3c0 .2-.2.4-.3.5l-.5.1h-16c-.6 0-1-.1-1.5-.5-.4-.3-.6-.9-.6-1.7V38.7c0-.2 0-.4.2-.5.1-.2.3-.3.6-.4l1-.2a15.7 15.7 0 0 1 4.6.2c.3.1.5.2.6.4l.1.5v28h11l.5.1.3.5.2 1v1.3M408.4 70c.2.6.4 1 .4 1.4 0 .4 0 .7-.2.9-.2.2-.5.3-1 .3a24.3 24.3 0 0 1-5.4 0 1 1 0 0 1-.6-.4 2 2 0 0 1-.3-.6l-2.3-7h-13.1l-2.2 6.8c0 .3-.2.5-.3.6-.1.2-.3.4-.6.4a26 26 0 0 1-5 .2c-.3 0-.7-.2-.8-.4-.2-.2-.3-.5-.2-.8 0-.4.2-1 .4-1.5l10.7-31 .4-.7c.1-.2.4-.3.7-.4l1.3-.2a47.7 47.7 0 0 1 6.2.2c.3 0 .6.2.7.4l.4.8 10.8 31zm-16-25.7l-5 14.9h10l-5-14.9zM22.5 68.5c-6.9-6.9-6.9-18 0-24.9l23.8-23.8a17.6 17.6 0 0 1 20-3.4v.1L42.9 39.9a5.9 5.9 0 0 0 8.2 8.3l4.2-4.1c2.3-2.4 4.7-2.4 6.7-.4l9 8.9c2.3 2.4 2.6 4.4 0 7.1l-8.5 8.5c-2.4 2.3-4.6 2.5-7.1 0L39 51.9a5.9 5.9 0 0 0-8.3 8.3L51.5 81c4 4 10.6 4 14.7 0l17.4-17.5c4-4 4-10.6 0-14.7L70.3 35.4c-2.1-2-2-4.4.2-6.6l4.3-4.2.5-.7L92 40.6a21.8 21.8 0 0 1 0 31L74.2 89.2a21.9 21.9 0 0 1-30.9 0L22.5 68.5z",
  },
  sv: {
    viewBox: "0 0 463 111",
    title: "LokalTapiola",
    path: "M131 69.6l-.1 1.4-.2.9-.4.5-.4.2h-16c-.6 0-1.1-.2-1.5-.6-.4-.3-.6-.9-.6-1.7V38.7l.1-.5.6-.4 1.1-.2a15 15 0 0 1 4.5.2c.3.1.5.2.6.4l.2.5v28h11c.1 0 .3 0 .4.2.2 0 .3.2.4.4l.2 1v1.3M157.4 55c0-1.7-.1-3.4-.4-4.9-.3-1.5-.8-2.8-1.5-3.9a7 7 0 0 0-2.7-2.4 9.5 9.5 0 0 0-4.3-1c-1.8 0-3.2.4-4.4 1a8 8 0 0 0-2.8 2.7c-.8 1-1.3 2.4-1.6 3.8a27.6 27.6 0 0 0 0 9.8 11 11 0 0 0 1.5 4c.7 1 1.6 1.8 2.7 2.4 1.2.6 2.6.8 4.4.8a7.9 7.9 0 0 0 7.2-3.6c.7-1 1.2-2.4 1.5-3.9.3-1.5.4-3 .4-4.7m7.5-.4c0 3-.4 5.5-1.1 7.8-.7 2.3-1.8 4.2-3.2 5.8a13.7 13.7 0 0 1-5.3 3.6c-2 .9-4.5 1.3-7.3 1.3-2.7 0-5-.4-7-1a12.4 12.4 0 0 1-8.1-8.9 28 28 0 0 1 0-15.6 14 14 0 0 1 8.5-9.4c2.1-.8 4.5-1.2 7.3-1.2 2.6 0 5 .3 7 1a12.4 12.4 0 0 1 8.1 8.8c.7 2.2 1 4.8 1 7.8M196.3 71.5l-.1.5a1 1 0 0 1-.6.4l-1.2.2a17 17 0 0 1-4.3 0c-.5-.1-.9-.2-1.1-.4l-.5-.7L177.9 55v16.5l-.1.6-.6.3-1 .2a15 15 0 0 1-4.6-.2c-.3 0-.5-.2-.6-.3a.8.8 0 0 1-.2-.6V38.7c0-.2 0-.4.2-.5.1-.2.3-.3.6-.4l1-.2a14.8 14.8 0 0 1 4.6.2c.3.1.5.2.6.4l.1.5v14.8l10.2-14.8.5-.6.7-.3a20.7 20.7 0 0 1 5.7 0l.6.4.2.5-.3 1-.9 1.5-9.6 12.5 10.5 15.8.7 1.4.1.6M213.6 44.3l-5 14.9h10l-5-14.9zm16 25.6l.4 1.6c0 .3 0 .6-.2.8-.2.2-.6.3-1 .4a26.7 26.7 0 0 1-5.5-.2 1 1 0 0 1-.5-.3l-.3-.6-2.3-7H207l-2.2 6.8-.3.7-.6.4-1.1.2a27.7 27.7 0 0 1-3.8 0c-.5-.1-.8-.2-1-.4-.2-.2-.2-.5-.2-1l.4-1.4 10.8-31c0-.3.2-.5.4-.7.1-.2.3-.4.6-.4l1.4-.2a44.9 44.9 0 0 1 6.2.2c.3 0 .6.2.7.4l.4.8 10.8 31zM253.6 69.6V71l-.2.9-.4.5-.4.2h-16c-.6 0-1.1-.2-1.5-.6-.4-.3-.6-.9-.6-1.7V38.7c0-.2 0-.4.2-.5 0-.2.3-.3.5-.4l1.1-.2a15 15 0 0 1 4.5.2c.3.1.5.2.6.4l.2.5v28h11c.1 0 .3 0 .4.2.2 0 .3.2.4.4l.2 1v1.3M276.3 40.6v1.3l-.3 1-.3.4-.5.2h-8.7v28.1c0 .2 0 .3-.2.5l-.6.3-1 .2a15.2 15.2 0 0 1-4.6-.2l-.6-.3a.7.7 0 0 1-.1-.5V43.5h-8.8c-.1 0-.3 0-.4-.2-.2 0-.3-.3-.3-.5l-.2-.9a10.7 10.7 0 0 1 .2-3.6c0-.2.1-.4.3-.5l.4-.1h24.6l.5.1.3.5.2 1v1.3M288.6 44.3l-5 14.9h10l-5-14.9zm16 25.6l.4 1.6c0 .3 0 .6-.2.8-.2.2-.5.3-1 .4a26.8 26.8 0 0 1-5.4-.2 1 1 0 0 1-.6-.3l-.3-.6-2.3-7H282l-2.2 6.8-.3.7-.6.4-1 .2a27.9 27.9 0 0 1-3.9 0c-.5-.1-.8-.2-1-.4-.2-.2-.2-.5-.2-1l.4-1.4 10.8-31 .4-.7c.1-.2.3-.4.7-.4l1.3-.2a45 45 0 0 1 6.2.2c.3 0 .6.2.7.4l.4.8 10.8 31zM326 48.8c0-1.2-.3-2.2-.7-3-.4-.8-1-1.4-1.6-1.8-.6-.4-1.3-.6-2-.7l-2.1-.1h-3V55h3.2c1.1 0 2-.2 2.8-.5.8-.3 1.4-.7 1.9-1.2a8 8 0 0 0 1.5-4.5m7.4-.5c0 2-.3 3.7-1 5.2a10 10 0 0 1-2.6 3.8c-1.2 1-2.6 1.8-4.3 2.4a20 20 0 0 1-6 .8h-3V72l-.7.3-1 .2a14.8 14.8 0 0 1-4.6-.2c-.2 0-.4-.2-.5-.3a.8.8 0 0 1-.2-.5V40.2c0-.8.2-1.5.7-1.9.4-.4 1-.6 1.7-.6h8.2a27.3 27.3 0 0 1 5 .5c1.1.2 2.2.6 3.2 1.1 1 .6 2 1.3 2.8 2.1.7.9 1.3 1.9 1.7 3 .4 1.2.6 2.5.6 3.9M346 71.6c0 .2 0 .3-.2.5l-.6.3-1 .2a15.1 15.1 0 0 1-4.6-.2c-.2 0-.4-.2-.5-.3a.8.8 0 0 1-.2-.5v-33c0-.1 0-.3.2-.4 0-.2.3-.3.5-.4l1.1-.2a14.5 14.5 0 0 1 4.5.2c.3.1.5.2.6.4l.2.5v32.9zM377.5 55c0-1.7-.1-3.4-.4-4.9-.3-1.5-.8-2.8-1.5-3.9a7 7 0 0 0-2.7-2.4 9.5 9.5 0 0 0-4.3-1c-1.8 0-3.2.4-4.4 1a8 8 0 0 0-2.8 2.7c-.8 1-1.3 2.4-1.6 3.8a27.8 27.8 0 0 0 0 9.8 11 11 0 0 0 1.5 4c.7 1 1.6 1.8 2.7 2.4 1.2.6 2.6.8 4.4.8a7.9 7.9 0 0 0 7.2-3.6c.7-1 1.2-2.4 1.5-3.9.3-1.5.4-3 .4-4.7m7.5-.4c0 3-.4 5.5-1.1 7.8-.7 2.3-1.8 4.2-3.2 5.8a13.8 13.8 0 0 1-5.3 3.6c-2 .9-4.5 1.3-7.2 1.3-2.8 0-5.1-.4-7.1-1a12.3 12.3 0 0 1-8.1-8.9 28 28 0 0 1 0-15.6 14 14 0 0 1 8.5-9.4c2.1-.8 4.5-1.2 7.3-1.2 2.6 0 5 .3 7 1a12.4 12.4 0 0 1 8.1 8.8c.7 2.2 1 4.8 1 7.8M410.1 69.6V71l-.3.9c0 .2-.2.4-.3.5l-.5.2h-16c-.6 0-1-.2-1.5-.6-.4-.3-.6-.9-.6-1.7V38.7c0-.2 0-.4.2-.5.1-.2.3-.3.6-.4l1-.2a14.9 14.9 0 0 1 4.6.2c.3.1.5.2.6.4l.2.5v28h11c.1 0 .3 0 .4.2l.3.4.2 1 .1 1.3M426.8 44.3l-5 14.9h10l-5-14.9zm16 25.6l.5 1.6c0 .3 0 .6-.3.8-.2.2-.5.3-1 .4a26.8 26.8 0 0 1-5.4-.2 1 1 0 0 1-.6-.3l-.2-.6-2.4-7h-13l-2.3 6.8c0 .3-.2.5-.3.7l-.5.4-1.1.2a27.6 27.6 0 0 1-3.9 0l-1-.4c-.1-.2-.2-.5-.1-1l.4-1.4 10.7-31 .4-.7c.2-.2.4-.4.7-.4l1.3-.2a45 45 0 0 1 6.2.2c.4 0 .6.2.8.4l.4.8 10.8 31zM21.8 68.6c-7-7-7-18.1 0-25l23.8-23.8a17.7 17.7 0 0 1 20-3.4L42.2 40a5.9 5.9 0 0 0 8.2 8.3l4.2-4.1c2.3-2.4 4.6-2.4 6.7-.4l9 9c2.3 2.3 2.6 4.4 0 7l-8.6 8.5c-2.3 2.3-4.5 2.5-7 0L38.3 51.9a5.9 5.9 0 0 0-8.3 8.3L50.8 81c4 4 10.6 4 14.6 0L83 63.4c4-4 4-10.6 0-14.7L69.6 35.4c-2.1-2-2-4.4.2-6.6l4.3-4.3.5-.6 16.7 16.7a21.9 21.9 0 0 1 0 31L73.5 89.3a21.9 21.9 0 0 1-31 0L21.9 68.6z",
  },
  en: {
    viewBox: "0 0 460 111",
    title: "LocalTapiola",
    path: "M131.6 69.7V71l-.3 1c0 .2-.2.4-.3.5l-.5.1h-16c-.5 0-1-.1-1.4-.5-.5-.3-.7-1-.7-1.7V38.7l.2-.4c.1-.2.3-.3.6-.4l1-.2a14.8 14.8 0 0 1 4.6.2c.3.1.5.2.6.4l.2.4v28h11l.4.2.3.5.2 1 .1 1.3M158 55.2c0-1.8 0-3.5-.4-5-.2-1.5-.7-2.8-1.4-3.9a7 7 0 0 0-2.8-2.4 9.4 9.4 0 0 0-4.3-1c-1.7 0-3.2.4-4.4 1a8 8 0 0 0-2.8 2.7c-.7 1-1.2 2.4-1.5 3.8a27.6 27.6 0 0 0 0 9.8 11 11 0 0 0 1.4 3.9c.7 1 1.6 1.9 2.8 2.5 1.1.5 2.6.8 4.3.8a7.9 7.9 0 0 0 7.2-3.6c.7-1.1 1.2-2.4 1.5-3.9.3-1.5.5-3 .5-4.7m7.4-.4c0 2.9-.4 5.5-1 7.8-.8 2.2-1.9 4.2-3.3 5.8A13.7 13.7 0 0 1 156 72c-2.1.8-4.6 1.3-7.3 1.3s-5-.4-7.1-1.1a12.3 12.3 0 0 1-8-8.8 28 28 0 0 1 0-15.6 14 14 0 0 1 8.5-9.4c2-.8 4.5-1.2 7.2-1.2s5 .3 7 1 3.7 1.8 5.1 3.2c1.4 1.5 2.4 3.3 3 5.6.8 2.2 1.1 4.8 1.1 7.8M213.2 44.4l-5 14.8h10l-5-14.8zm16 25.6c.3.6.4 1.2.4 1.5 0 .4 0 .7-.2.9-.2.2-.5.3-1 .3a26.6 26.6 0 0 1-5.4-.1c-.3 0-.5-.2-.6-.3l-.3-.6-2.3-7h-13l-2.3 6.8c0 .3-.2.5-.3.6l-.5.4-1.1.2a27.2 27.2 0 0 1-3.9 0c-.4 0-.8-.2-1-.4-.1-.2-.2-.5-.1-.8l.4-1.6L208.7 39l.4-.7.7-.5 1.3-.1a47 47 0 0 1 6.2.1c.4.1.6.3.8.5l.4.8L229.2 70zM253.3 69.7l-.1 1.3-.2 1-.3.5-.5.1h-16c-.6 0-1-.1-1.5-.5-.4-.3-.6-1-.6-1.7V38.7c0-.1 0-.3.2-.4.1-.2.3-.3.6-.4l1-.2a14.8 14.8 0 0 1 4.6.2l.5.4c.2.1.2.3.2.4v28h11l.5.2.3.5.2 1v1.3M275.9 40.7V42l-.3 1-.3.4-.5.2h-8.7v28c0 .2 0 .4-.2.6l-.6.3-1 .2a15.3 15.3 0 0 1-4.6-.2c-.2 0-.4-.2-.5-.3a.7.7 0 0 1-.2-.5V43.6h-8.7c-.2 0-.4 0-.5-.2l-.3-.5-.2-.9a10.9 10.9 0 0 1 .2-3.6c0-.2.2-.4.3-.5l.5-.1h24.5l.5.1.3.5.2 1v1.3M288.2 44.4l-5 14.8h10l-5-14.8zm16 25.6l.4 1.5c0 .4 0 .7-.2.9-.2.2-.6.3-1 .3a26.7 26.7 0 0 1-5.5-.1 1 1 0 0 1-.5-.3l-.3-.6-2.3-7h-13.1l-2.2 6.8-.3.6c-.1.2-.3.3-.6.4l-1 .2a27.3 27.3 0 0 1-3.9 0c-.5 0-.8-.2-1-.4-.2-.2-.2-.5-.2-.8l.4-1.6L283.7 39l.4-.7c.1-.2.3-.4.7-.5l1.3-.1a47.4 47.4 0 0 1 6.2.1c.3.1.6.3.7.5l.4.8L304.2 70zM325.5 48.9c0-1.2-.2-2.2-.6-3-.5-.8-1-1.4-1.6-1.8-.6-.4-1.3-.6-2-.7l-2.1-.2h-3v11.9h3.1c1.2 0 2.1-.2 2.9-.5.7-.3 1.4-.7 1.8-1.3.5-.5.9-1.2 1.1-2 .3-.7.4-1.5.4-2.4m7.4-.5c0 2-.3 3.7-.9 5.2a10 10 0 0 1-2.6 3.8c-1.2 1-2.6 1.8-4.3 2.3-1.7.6-3.7.8-6 .8h-3v11.2l-.1.5-.6.3-1 .2a14.9 14.9 0 0 1-4.6-.2c-.3 0-.5-.2-.6-.3a.8.8 0 0 1-.1-.5V40.3c0-.8.2-1.5.6-1.9.5-.4 1-.6 1.8-.6h8.2a27.6 27.6 0 0 1 5 .5c1 .2 2.1.6 3.2 1.1a8.9 8.9 0 0 1 4.5 5.1c.3 1.2.5 2.4.5 3.9M345.5 71.7l-.1.5-.6.3-1 .2a15 15 0 0 1-4.6-.2c-.3 0-.5-.2-.6-.3a.8.8 0 0 1-.2-.5v-33c0-.1 0-.3.2-.4.1-.2.3-.3.6-.4l1-.2a14.5 14.5 0 0 1 4.6.2c.3.1.5.2.6.4l.1.4v33zM377 55.2c0-1.8 0-3.5-.4-5-.3-1.5-.7-2.8-1.4-3.9a7 7 0 0 0-2.8-2.4 9.4 9.4 0 0 0-4.3-1c-1.8 0-3.2.4-4.4 1a8 8 0 0 0-2.8 2.7c-.7 1-1.2 2.4-1.5 3.8a27.8 27.8 0 0 0 0 9.8 11 11 0 0 0 1.4 3.9c.7 1 1.6 1.9 2.8 2.5 1.1.5 2.6.8 4.3.8a7.9 7.9 0 0 0 7.2-3.6c.7-1.1 1.2-2.4 1.5-3.9.3-1.5.5-3 .5-4.7m7.4-.4c0 2.9-.4 5.5-1 7.8-.8 2.2-1.9 4.2-3.3 5.8A13.8 13.8 0 0 1 375 72a21.6 21.6 0 0 1-14.4.2 12.3 12.3 0 0 1-8-8.8 28 28 0 0 1 0-15.6c.8-2.2 1.9-4.2 3.3-5.7a14 14 0 0 1 5.2-3.7c2.1-.8 4.6-1.2 7.3-1.2s5 .3 7 1 3.7 1.8 5 3.2c1.4 1.5 2.5 3.3 3.2 5.6.7 2.2 1 4.8 1 7.8M409.6 69.7a11.5 11.5 0 0 1-.3 2.3c0 .2-.2.4-.3.5l-.5.1h-16c-.5 0-1-.1-1.4-.5-.4-.3-.7-1-.7-1.7V38.7l.2-.4c.1-.2.3-.3.6-.4l1.1-.2a14.7 14.7 0 0 1 4.5.2c.3.1.5.2.6.4l.2.4v28h11l.4.2.3.5.2 1 .1 1.3M426.3 44.4l-5 14.8h10l-5-14.8zm16 25.6c.3.6.4 1.2.4 1.5.1.4 0 .7-.2.9-.2.2-.5.3-1 .3a26.7 26.7 0 0 1-5.4-.1c-.3 0-.5-.2-.6-.3l-.3-.6-2.3-7h-13l-2.3 6.8c0 .3-.2.5-.3.6 0 .2-.3.3-.5.4l-1.1.2a27.1 27.1 0 0 1-3.9 0c-.4 0-.7-.2-1-.4-.1-.2-.2-.5-.1-.8l.4-1.6L421.8 39l.4-.7.7-.5 1.3-.1a47.3 47.3 0 0 1 6.2.1c.4.1.6.3.8.5l.4.8L442.3 70zM22.5 68.6c-6.9-6.9-6.9-18 0-25L46.3 20a17.6 17.6 0 0 1 20-3.4v.1L42.9 40a5.9 5.9 0 0 0 8.2 8.3l4.2-4.1c2.3-2.4 4.7-2.4 6.7-.4l9 8.9c2.3 2.4 2.6 4.4 0 7.1l-8.5 8.5c-2.4 2.3-4.6 2.5-7.1 0L39 52a5.9 5.9 0 0 0-8.3 8.3L51.5 81c4 4 10.6 4 14.7 0l17.4-17.5c4-4 4-10.6 0-14.7L70.3 35.5c-2.1-2-2-4.4.2-6.6l4.3-4.3.5-.6L92 40.7a21.8 21.8 0 0 1 0 31L74.2 89.3a21.8 21.8 0 0 1-30.9 0L22.5 68.6zM195.7 67.4a13.7 13.7 0 0 1-.2 2l-.2.5-.5.6c-.2.2-.5.5-1.1.8l-2 1a18.2 18.2 0 0 1-6.3 1A18 18 0 0 1 179 72c-2-.7-3.6-1.8-5-3.3a14.7 14.7 0 0 1-3-5.4c-.7-2.2-1.1-4.8-1.1-7.7 0-3 .4-5.6 1.2-7.9.8-2.3 1.9-4.2 3.3-5.7 1.4-1.6 3.1-2.8 5-3.6a17.2 17.2 0 0 1 14 .6l1.3.8.5.6.2.7.1.9a16.4 16.4 0 0 1-.2 3.5c0 .2-.2.4-.3.5l-.5.2-1-.5a16.8 16.8 0 0 0-4-2 10 10 0 0 0-6.9.3 8 8 0 0 0-2.7 2.4c-.8 1-1.4 2.3-1.8 3.8-.4 1.5-.6 3.2-.6 5 0 2 .3 3.9.7 5.4.4 1.5 1 2.7 1.8 3.7s1.7 1.7 2.8 2.1c1.1.5 2.3.7 3.7.7a10.6 10.6 0 0 0 5.6-1.4l1.7-1c.4-.2.7-.4 1-.4l.4.1c0 .1.2.2.2.5l.2.9v1.5",
  },
};
const DefaultLogo = ({ inverse, logoId, language }) => {
  const logoColor = inverse ? colorGrayLightest : colorPrimary;
  const { title, path, viewBox } = i18n[language];
  return (h("svg", { role: "img", xmlns: "http://www.w3.org/2000/svg", fill: logoColor, "aria-labelledby": logoId, viewBox: viewBox }, h("title", { id: logoId }, title), h("path", { d: path })));
};
export class DuetLogo {
  constructor() {
    /**
     * Own Properties.
     */
    this.logoId = createID("DuetLogo");
    this.theme = "";
    this.margin = "auto";
    this.inverse = false;
    this.size = "medium";
    this.href = "/";
    this.accessibleLabelDefaults = {
      fi: "Siirry etusivulle.",
      sv: "Gå till förstasidan.",
      en: "Go to front page.",
    };
    this.accessibleLabel = getLocaleString(this.accessibleLabelDefaults);
    this.language = getLanguage();
  }
  watchLanguageStateHandler(newValue, oldValue) {
    if (newValue !== oldValue && this.accessibleLabel === this.accessibleLabelDefaults[oldValue]) {
      this.accessibleLabel = getLocaleString(this.accessibleLabelDefaults, newValue);
    }
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  connectedCallback() {
    connectLanguageChangeObserver(this);
  }
  disconnectedCallback() {
    disconnectLanguageChangeObserver(this);
  }
  /**
   * Sets focus on the specified `duet-logo`. Use this method instead of the global
   * `logo.focus()`.
   */
  async setFocus(options) {
    if (this.nativeLink) {
      this.nativeLink.focus(options);
    }
  }
  /**
   * render() function
   * Always the last one in the class.
   */
  render() {
    const TagName = this.href ? "a" : "div";
    return (h(Host, { class: { "duet-m-0": this.margin === "none" } }, h(TagName, { href: this.href, ref: link => (this.nativeLink = link), "aria-label": this.accessibleLabel, class: {
        "duet-logo": true,
        [this.size]: true,
        "duet-theme-turva": this.theme === "turva",
        "duet-logo-en": this.language === "en",
        "duet-logo-sv": this.language === "sv",
      } }, this.theme === "turva" ? (h(TurvaLogo, { logoId: this.logoId, inverse: this.inverse })) : (h(DefaultLogo, { logoId: this.logoId, inverse: this.inverse, language: this.language })))));
  }
  static get is() { return "duet-logo"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["duet-logo.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["duet-logo.css"]
    };
  }
  static get properties() {
    return {
      "theme": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetTheme",
          "resolved": "\"\" | \"default\" | \"turva\"",
          "references": {
            "DuetTheme": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Theme of the logo."
        },
        "attribute": "theme",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "margin": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetMargin",
          "resolved": "\"auto\" | \"none\"",
          "references": {
            "DuetMargin": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Controls the margin of the component."
        },
        "attribute": "margin",
        "reflect": false,
        "defaultValue": "\"auto\""
      },
      "inverse": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Use inverse version of the logo (white logo on dark background)."
        },
        "attribute": "inverse",
        "reflect": false,
        "defaultValue": "false"
      },
      "size": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetLogoSize",
          "resolved": "\"large\" | \"medium\"",
          "references": {
            "DuetLogoSize": {
              "location": "local",
              "path": "/Users/laaksju3/duetds/packages/components/src/components/duet-logo/duet-logo.tsx"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Size variation of the logo."
        },
        "attribute": "size",
        "reflect": false,
        "defaultValue": "\"medium\""
      },
      "href": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Where the logo links to."
        },
        "attribute": "href",
        "reflect": true,
        "defaultValue": "\"/\""
      },
      "accessibleLabelDefaults": {
        "type": "unknown",
        "mutable": true,
        "complexType": {
          "original": "DuetLangObject",
          "resolved": "DuetLangObject",
          "references": {
            "DuetLangObject": {
              "location": "import",
              "path": "../../utils/language-utils"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Default translations for accessible label."
        },
        "defaultValue": "{\n    fi: \"Siirry etusivulle.\",\n    sv: \"G\u00E5 till f\u00F6rstasidan.\",\n    en: \"Go to front page.\",\n  }"
      },
      "accessibleLabel": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Accessible label. Defaults to what is in accessibleLabelDefaults prop."
        },
        "attribute": "accessible-label",
        "reflect": true,
        "defaultValue": "getLocaleString(this.accessibleLabelDefaults)"
      },
      "language": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetLanguage",
          "resolved": "\"en\" | \"fi\" | \"sv\"",
          "references": {
            "DuetLanguage": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "deprecated",
              "text": "this is now handled via the html lang tag, and is no longer used - kept to avoid breaking changes and ease unit testing"
            }, {
              "name": "default",
              "text": "\"fi\""
            }],
          "text": "The language of the logo."
        },
        "attribute": "language",
        "reflect": true,
        "defaultValue": "getLanguage()"
      }
    };
  }
  static get methods() {
    return {
      "setFocus": {
        "complexType": {
          "signature": "(options?: FocusOptions) => Promise<void>",
          "parameters": [{
              "tags": [],
              "text": ""
            }],
          "references": {
            "Promise": {
              "location": "global"
            },
            "FocusOptions": {
              "location": "global"
            }
          },
          "return": "Promise<void>"
        },
        "docs": {
          "text": "Sets focus on the specified `duet-logo`. Use this method instead of the global\n`logo.focus()`.",
          "tags": []
        }
      }
    };
  }
  static get elementRef() { return "element"; }
  static get watchers() {
    return [{
        "propName": "language",
        "methodName": "watchLanguageStateHandler"
      }];
  }
}
