// @duetds/components/lib/collection/components/duet-choice/duet-checkmark.js
/*!
 * Built with Duet Design System
 */
import { h, Host } from "@stencil/core";
import { inheritGlobalTheme } from "../../common";
export class DuetCheckmark {
  constructor() {
    this.theme = "";
    this.checked = false;
    this.presentationOnly = false;
    this.center = false;
    this.type = "checkbox";
  }
  /**
   * Component lifecycle events.
   */
  componentWillLoad() {
    inheritGlobalTheme(this);
  }
  render() {
    return (h(Host, { class: { "duet-theme-turva": this.theme === "turva" } }, h("div", { role: this.presentationOnly ? "presentation" : null, class: { checked: this.checked } }, h("div", { class: {
        "duet-checkmark-container": true,
        center: this.center,
        "duet-checkmark-radio": this.type === "radio",
      } }, this.checked && h("div", { class: "duet-checkmark" })))));
  }
  static get is() { return "duet-checkmark"; }
  static get originalStyleUrls() {
    return {
      "$": ["duet-checkmark.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["duet-checkmark.css"]
    };
  }
  static get properties() {
    return {
      "theme": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "DuetTheme",
          "resolved": "\"\" | \"default\" | \"turva\"",
          "references": {
            "DuetTheme": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Theme"
        },
        "attribute": "theme",
        "reflect": false,
        "defaultValue": "\"\""
      },
      "checked": {
        "type": "boolean",
        "mutable": true,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Checked"
        },
        "attribute": "checked",
        "reflect": false,
        "defaultValue": "false"
      },
      "presentationOnly": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Checked"
        },
        "attribute": "presentation-only",
        "reflect": false,
        "defaultValue": "false"
      },
      "center": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Checked"
        },
        "attribute": "center",
        "reflect": false,
        "defaultValue": "false"
      },
      "type": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "DuetChoiceType",
          "resolved": "\"checkbox\" | \"radio\"",
          "references": {
            "DuetChoiceType": {
              "location": "import",
              "path": "../../common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Type"
        },
        "attribute": "type",
        "reflect": true,
        "defaultValue": "\"checkbox\""
      }
    };
  }
}
